---
title: "St. John symbiont time series"
author: "R. Cunning"
date: "2/16/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r libraries}

library(janitor)
library(phyloseq)  # BiocManager::install("phyloseq")
library(cowplot)
library(ggrepel)
library(scales)
library(RColorBrewer)
library(MASS)
library(lme4)
library(emmeans)
library(tidyverse)
library(vegan)
library(ggh4x)

# library(conflicted)
# conflict_prefer("select", "dplyr")
# conflict_prefer("filter", "dplyr")
# conflict_prefer("rename", "dplyr")
```

# Load Data
```{r profiles, include = FALSE}
sam0 <- read_csv("Data/STJ_Metadata_2017_2019.csv") %>%
  clean_names()
sam1 <- as.matrix(sam0)
rownames(sam1) <- sam0$sample_name
sam <- sample_data(data.frame(sam1))

tax0 <- read_tsv(
  file = "Data/20200623_lenz/its2_type_profiles/111_20200624_2020-06-24_03-40-50.238182.profiles.absolute.abund_and_meta.txt",
  n_max = 6) %>%
  dplyr::select(-2) %>% 
  gather(UID, value, -1) %>% 
  spread(1, value) %>%
  clean_names()
tax1 <- as.matrix(tax0[, -1], dimnames = list(tax0$uid, colnames(tax0[-1])))
rownames(tax1) <- tax0$uid
tax <- tax_table(tax1)

otu0 <- read_tsv(
  file = "Data/20200623_lenz/its2_type_profiles/111_20200624_2020-06-24_03-40-50.238182.profiles.absolute.abund_and_meta.txt") %>% 
  rename(sample_name = X2) %>%
  select(-1) %>%
  slice(7:n()) %>%
  mutate_at(2:ncol(.), as.numeric)
otu1 <- as.matrix(otu0[, -1])
rownames(otu1) <- otu0$sample_name
otu <- otu_table(otu1, taxa_are_rows = FALSE)

profiles <- phyloseq(otu, tax, sam)
```

```{r variants, include = FALSE}
taxnames <- read_tsv(
  file = "Data/20200623_lenz/post_med_seqs/111_20200624_2020-06-24_03-40-50.238182.seqs.absolute.abund_and_meta.txt",
  n_max = 0) %>%
  select(-(1:33)) %>%
  names(.)
tax0 <- data_frame(
  DIV = taxnames,
  clade = str_extract(DIV, "[A-Z]")
)
tax1 <- as.matrix(tax0)
rownames(tax1) <- tax0$DIV
tax <- tax_table(tax1)

otu0 <- read_tsv(
  file = "Data/20200623_lenz/post_med_seqs/111_20200624_2020-06-24_03-40-50.238182.seqs.absolute.abund_and_meta.txt") %>% 
  select(-1, -(3:33))
otu1 <- as.matrix(otu0[, -1])
rownames(otu1) <- otu0$sample_name
otu <- otu_table(otu1, taxa_are_rows = FALSE)

variants <- phyloseq(otu, tax, sam)

tax_table(variants)
```

```{r filter_save, include = FALSE}
# Filter out RANDOM samples (non-tagged colonies)
variants <- subset_samples(variants, host_id != "RANDOM")
variants <- prune_taxa(taxa_sums(variants) != 0, variants)
profiles <- subset_samples(profiles, host_id != "RANDOM")
profiles <- prune_taxa(taxa_sums(profiles) != 0, profiles)

save(profiles, variants, file = "data/Symb_phyloseq.RData")
```

# Create custom functions, color palettes, and ggplot themes
```{r color_palettes, include = FALSE}
# Create standardized color pallettes for plotting profiles and variants

#ppal <- sample(allcolors, ntaxa(profiles))
#ppal <- hue_pal(l = 65)(ntaxa(profiles))
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
set.seed(555)
ppal <- sample(col_vector, ntaxa(profiles))
ppal <- setNames(ppal, data.frame(tax_table(profiles))$its2_type_profile)

allcolors <- grDevices::colors()
set.seed(888)
vpal <- sample(allcolors, ntaxa(variants))
#vpal <- hue_pal(l = 65)(ntaxa(variants))
vpal <- setNames(vpal, taxa_names(variants))



# Coral colors

```

```{r plotting_functions, include = FALSE}
# Plotting functions
plot_both <- function(mycol) {
  s <- subset(sample_data(variants), host_id == mycol)
  v <- merge_phyloseq(tax_table(variants), otu_table(variants), s)
  p <- merge_phyloseq(tax_table(profiles), otu_table(profiles), s)
  v <- prune_taxa(taxa_sums(v) != 0, v)
  p <- prune_taxa(taxa_sums(p) != 0, p)
  vmdf <- psmelt(v)
  seqdepth <- vmdf %>%
    group_by(year) %>%
    summarize(sum = paste0("n=", sum(Abundance)))
  ti <- paste0(vmdf[1, c("host_genus", "host_species", "host_id", "host_site")], collapse = " ")
  v_plot <- ggplot(vmdf, aes(x = year, y = Abundance, fill = DIV, label = DIV)) + 
    geom_bar(stat = "identity", position = "fill", color = "black", lwd = 0.1) + 
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)), labels = scales::percent) +
    scale_fill_manual(values = vpal[vmdf$OTU]) +
    geom_text(aes(size = Abundance), position = position_fill(vjust = 0.5)) +
    geom_text(data = seqdepth, aes(x = year, y = 1.01, label = sum), vjust = 0, size = 3, inherit.aes = FALSE) +
    theme(legend.position = "none", panel.background = element_blank(), 
          axis.title.x = element_blank(), axis.title.y = element_blank()) +
    coord_cartesian(ylim = c(0, 1))
  pmdf <- psmelt(p)
  p_plot <- ggplot(pmdf, aes(x = year, y = Abundance, fill = its2_type_profile, label = its2_type_profile)) +
    geom_bar(stat = "identity", position = "fill", color = "black", lwd = 0.1) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)), labels = scales::percent) +
    scale_fill_manual(values = ppal[pmdf$its2_type_profile]) +
    geom_label_repel(position = position_fill(vjust = 0.5), force = 2,
                     point.size = NA, xlim = c(-0.1, 3.6), ylim = c(0.1,0.9), size = 2) +
    theme(legend.position = "none", panel.background = element_blank(), 
          axis.title.x = element_blank(), axis.title.y = element_blank()) +
    coord_cartesian(clip = "off")
  plot_row <- plot_grid(p_plot, v_plot, labels = c("ITS2 Profiles", "ITS2 Variants"), label_fontface = "plain")
  title <- ggdraw() + 
    draw_label(ti, fontface = 'bold', x = 0, hjust = 0) +
    theme(plot.margin = margin(0, 0, 0, 7))
  plot_grid(title, plot_row, ncol = 1, rel_heights = c(0.1, 1))
}

#plot_both(mycol = all3[1])

# Plot by species and site
# ordf <- data.frame(sam) %>%
#   distinct(host_id, host_genus, host_species, host_site) %>%
#   filter(host_id != "RANDOM") %>%
#   filter(host_id %in% all3) %>%
#   arrange(host_genus, host_species, host_site)



# new plotting function to condense by species rather than by individual
plot_both2 <- function(...) {
  s <- subset(sample_data(variants), ...)
  v <- merge_phyloseq(tax_table(variants), otu_table(variants), s)
  p <- merge_phyloseq(tax_table(profiles), otu_table(profiles), s)
  v <- prune_taxa(taxa_sums(v) != 0, v)
  p <- prune_taxa(taxa_sums(p) != 0, p)
  vmdf <- psmelt(v)
  seqdepth <- vmdf %>%
    group_by(year) %>%
    summarize(sum = paste0("n=", sum(Abundance)))
  ti <- paste0(vmdf[1, c("host_genus", "host_species")], collapse = " ")
  v_plot <- ggplot(vmdf, aes(x = year, y = Abundance, fill = DIV, label = DIV)) + 
    geom_bar(stat = "identity", position = "fill") + 
    facet_wrap(~ host_site + host_id, nrow = 1) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)), labels = scales::percent) +
    ylab("ITS2 sequences") +
    scale_fill_manual(values = vpal[vmdf$OTU]) +
    # geom_text(aes(size = Abundance), position = position_fill(vjust = 0.5)) +
    # geom_text(data = seqdepth, aes(x = year, y = 1.01, label = sum), vjust = 0, size = 3, inherit.aes = FALSE) +
    theme(legend.position = "none", panel.background = element_blank(), 
          axis.title.x = element_blank(), axis.text.x = element_blank(),
          axis.text.y = element_blank(), axis.ticks.y = element_blank()) +
    coord_cartesian(ylim = c(0, 1))
  pmdf <- psmelt(p)
  p_plot <- ggplot(pmdf, aes(x = year, y = Abundance, fill = its2_type_profile, label = its2_type_profile)) +
    geom_bar(stat = "identity", position = "fill") +
    facet_wrap(~ host_site + host_id, nrow = 1) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)), labels = scales::percent) +
    ylab("Profiles") +
    scale_fill_manual(values = ppal[pmdf$its2_type_profile]) +
    # geom_label_repel(position = position_fill(vjust = 0.5), force = 2,
    #                  point.size = NA, xlim = c(-0.1, 3.6), ylim = c(0.1,0.9), size = 2) +
    theme(legend.position = "bottom", panel.background = element_blank(), 
          axis.title.x = element_blank(), axis.text.x = element_text(angle = 90),
          axis.text.y = element_blank(), axis.ticks.y = element_blank(),
          strip.background = element_blank(), strip.text.x = element_blank())# +
    #guides(fill = guide_legend(nrow = 4)) # +
    #coord_cartesian(clip = "off")
  plot_grid(v_plot, p_plot, nrow = 2, label_fontface = "plain")
}
```

```{r ggplot_theme}
# Create custom ggplot theme
theme_custom <- function() {
  theme_bw(base_size = 10, base_family = "Arial") %+replace%
    theme(
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(), 
      panel.background = element_blank(),
      panel.border = element_rect(color = "black", fill = NA),
      legend.background = element_rect(fill = NA, colour = NA)
    )
}

# scale_color_corals <- function(aesthetic, ...){
#     ggplot2:::manual_scale(
#         aesthetic,
#         values = setNames(c('#00BFC4', '#CD9600', '#7CAE00', '#00BE67', 
#                             '#F8766D', '#00A9FF', '#C77CFF', '#FF61CC'),
#                           c('annularis', 'cavernosa', 'faveolata', 'franksi', 
#                             'labrynthiformis', 'natans', 'siderea', 'strigosa')), 
#         ...
#     )
# }

scale_color_corals <- function(aesthetic, ...) {
  ggplot2:::manual_scale(
    aesthetic,
    labels = c("M. cavernosa", "S. siderea", "O. annularis", 
               "D. labyrinthiformis", "O. franksi", "C. natans", 
               "P. strigosa", "P. furcata", "O. faveolata"),
    values = setNames(c("#00B9E3", "#00BA38", "#F8766D",
                        "#93AA00", "#FF61C3", "#00C19F",
                        "#619CFF", "#D39200", "#DB72FB"),
                      c("cavernosa", "siderea", "annularis",
                        "labrynthiformis", "franksi", "natans",
                        "strigosa", "furcata", "faveolata")),
    ...)
}

sppord <- factor(levels = c("Cnat", "Dlab", "Pstr", "Pfur",
                            "Ssid", "Oann", "Ofav", "Ofra", "Mcav"))

## ggplot labeller
sppnames <- c(
  Cnat = "C. natans",
  Dlab = "D. labyrinthiformis",
  Pstr = "P. strigosa",
  Pfur = "P. furcata",
  Ssid = "S. siderea",
  Oann = "O. annularis",
  Ofav = "O. faveolata",
  Ofra = "O. franksi",
  Mcav = "M. cavernosa"
)

sitenames <- c(
  EastTektite = "ET",
  WhitePoint = "WP",
  CabritteHorn = "CH"
)

global_labeller <- labeller(
  gspp = sppnames,
  host_site = sitenames
)
```

```{r DAtesting_function}
# DESEq2 differential abundance testing function
deseqDA <- function(host_ids) {
  colData = data.frame(sample_data(variants))
  colData$host_id <- factor(colData$host_id)
  dds <- DESeqDataSetFromMatrix(countData = t(data.frame(otu_table(variants))),
                                colData = colData,
                                design = ~ year)
  ## Subset colonies
  dds <- dds[, colData(dds)$host_id %in% host_ids]
  ### Remove genes counted zero times in subset
  dds <- dds[ rowSums(counts(dds)) > 0, ]
  ### Drop unused factor levels
  colData(dds) <- droplevels(colData(dds))
  # Remove genes with less than 1 mean count across samples
  dds <- dds[ rowMeans(counts(dds)) > 1, ]
  # Run DESeq pipeline
  design(dds) <- formula(~ host_id + year)
  dsr1 <- DESeq(dds)
  # Define contrasts of interest (differences between years)
  contr <- tibble(num = c("2018", "2019", "2019"),
                  den = c("2017", "2017", "2018"))
  # Get diff abund taxa for each contrast
  DA <- contr %>%
    mutate(dsr = pmap(list(num, den), ~results(dsr1, contrast = c("year", ..1, ..2))))
  DA <- DA %>%
    mutate(sig = map(dsr, ~ rownames_to_column(data.frame(.[which(.$padj < 0.1), ]), "gene"))) %>%
    unnest(sig)
  # get variance stabilized counts for plotting
  vsd <- varianceStabilizingTransformation(dds)
  vsd2 <- limma::removeBatchEffect(assay(vsd), vsd$host_id)
  taxcounts <- data.frame(vsd2[which(rownames(vsd2) %in% DA$gene), ])
  if(ncol(taxcounts) == 1) {
    taxcounts <- data.frame(t(taxcounts))
    rownames(taxcounts) <- unique(DA$gene)
    }
  taxcounts <- rownames_to_column(taxcounts, var = "variant") %>%
    pivot_longer(-variant, names_to = "sample_name", values_to = "count") %>%
    left_join(dplyr::select(data.frame(sample_data(variants)), "sample_name", "host_id", "year"))
}



```

# Table of all sampling done
```{r}
# Create table of number of tagged colonies sampled at each site at each year
table1 <- as_tibble(sample_data(profiles)) %>%
  drop_na(host_id) %>%
  group_by(year, host_species, host_site) %>%
  summarize(n = n()) %>%
  arrange(host_site, year) %>%
  pivot_wider(names_from = c(host_site, year), values_from = n) %>% 
  adorn_totals(where = "row")

write_csv(table1, path = "Output/Table1.csv")
```

# QC (filter out low read samples)
```{r QC, include = FALSE}
# Total number of post-MED sequences
sum(sample_sums(variants))
nsamples(variants)

# Filter out samples with low read counts
low <- sample_sums(variants) < 1000
table(low)  # Number of samples being filtered out due to low read counts
variants <- subset_samples(variants, !low)
variants <- prune_taxa(taxa_sums(variants) != 0, variants)
profiles <- subset_samples(profiles, !low)
profiles <- prune_taxa(taxa_sums(profiles) != 0, profiles)

# Read count analysis
hist(sample_sums(variants), breaks = 100)
# One sample is an outlier with very high read count -- remove for summary stats of remaining samples
outlier <- sample_sums(variants)[which(sample_sums(variants) == max(sample_sums(variants)))]
woout <- sample_sums(variants)[which(sample_sums(variants) != max(sample_sums(variants)))]
# Histogram of remaining samples read counts
hist(woout, breaks = 50)
# Mean and standard deviation
mean(woout); sd(woout)

# Number of sequence variants and profiles
ntaxa(variants)
ntaxa(profiles)

# Parse sequences and profiles by genus(/clade)
data.frame(tax_table(variants)) %>% as_tibble() %>%
  count(clade)
data.frame(tax_table(profiles)) %>% as_tibble() %>%
  count(clade)
```

# Barplots for whole dataset
```{r}
# Algal genotypes --- whole dataset summary figure

## Transform to relative abundance
v <- transform_sample_counts(variants, function(x) x/sum(x))     # Relative abundance
p <- transform_sample_counts(profiles, function(x) x/sum(x))     # Relative abundance

## Filter out variants less than 0.1% relative abundance for plotting
vt <- transform_sample_counts(v, function(x) ifelse(x > 0.001, x, 0))
vt <- prune_taxa(taxa_sums(vt) != 0, vt)

### Arrange colonies by similarity within species
mp <- merge_samples(vt, "host_id")
hc <- hclust(dist(data.frame(otu_table(mp))))
neword <- tibble(host_id = as.character(sample_data(mp)$host_id), ord = hc$order)


## Set colors for plotting variants, with distinct palettes by clade/genus
tt <- as_tibble(data.frame(tax_table(vt))) %>%
  arrange(clade, DIV)
nt <- tt %>% count(clade) %>% pivot_wider(names_from = clade, values_from = n)
acols <- colorRampPalette(brewer.pal(9, "Greys"))(nt$A)
bcols <- colorRampPalette(brewer.pal(9, "Greens"))(nt$B)
ccols <- colorRampPalette(brewer.pal(9, "Blues"))(nt$C)
dcols <- colorRampPalette(brewer.pal(9, "Oranges"))(nt$D)
set.seed(556)
vpal <- c(acols[shuffle(acols)], bcols[shuffle(bcols)], ccols[shuffle(ccols)], dcols[shuffle(dcols)])
vpal <- setNames(vpal, tt$DIV)

# Order data for plotting
divord <- as_tibble(data.frame(tax_table(vt))) %>%
  arrange(clade, DIV) %>%
  mutate(DIV = as_factor(DIV))

  
vmdf <- psmelt(vt) %>%
  left_join(neword) %>%
  arrange(host_genus, host_species, host_site, ord, year, clade, DIV) %>%
  mutate(host_id = as_factor(host_id),
         sample_id = as_factor(paste(host_genus, host_species, host_id, year, sep = "_")),
         genus_species = as_factor(paste(str_sub(host_genus, 1, 1), ". ", host_species, sep = "")),
         gspp = factor(paste0(str_sub(host_genus, 1, 1), str_sub(host_species, 1, 3)),
                       levels = levels(sppord)),
         DIV = factor(DIV, levels = divord$DIV))

# Plot variants
g1 <- ggplot(vmdf, aes(x = year, y = Abundance)) + 
  geom_col(width = 1, aes(fill = DIV)) + 
  facet_nested(~ gspp + host_site + host_id, 
               labeller = global_labeller,
               space = "free_x",
               nest_line = element_line(linetype = 1),
               remove_labels = "x") +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0), labels = scales::percent) +
  ylab("ITS2 sequences") +
  scale_fill_manual(values = vpal[vmdf$OTU]) +
  theme(legend.position = "none", panel.background = element_blank(),
        text = element_text(size = 6), plot.margin = margin(5.5, 5.5, 1, 5.5),
        strip.text.x = element_text(margin = margin(0.5, 0, 0.5, 0, "mm"), face = "italic"),
        axis.title.x = element_blank(), axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.spacing = unit(0.25, "mm"),
        strip.background = element_blank())
g1

# Repeat for profiles
### truncate its2_type_profile names
pp <- as_tibble(data.frame(tax_table(p))) %>%
  arrange(clade, its2_type_profile) %>%
  mutate(its2_type_profile_trunc = str_trunc(its2_type_profile, 22, "right"))
np <- pp %>% count(clade) %>% pivot_wider(names_from = clade, values_from = n)
acols <- tail(brewer.pal(5, "Greys"), np$A)
bcols <- colorRampPalette(brewer.pal(9, "YlGn"))(np$B)
ccols <- colorRampPalette(brewer.pal(9, "PuBu"))(np$C)
dcols <- tail(brewer.pal(9, "Reds"), np$D)
set.seed(556)
ppal <- c(acols, bcols, ccols, dcols)
ppal <- setNames(ppal, pp$its2_type_profile_trunc)

proford <- as_tibble(data.frame(tax_table(p))) %>%
  mutate(its2_type_profile_trunc = str_trunc(its2_type_profile, 22, "right")) %>%
  arrange(clade, its2_type_profile_trunc) %>%
  mutate(its2_type_profile = as_factor(its2_type_profile_trunc))
  
pmdf <- psmelt(p) %>%
  left_join(neword) %>%
  arrange(host_genus, host_species, host_site, ord, year, clade, its2_type_profile) %>%
  mutate(host_id = as_factor(host_id),
         its2_type_profile_trunc = str_trunc(its2_type_profile, 22, "right"),
         sample_id = as_factor(paste(host_genus, host_species, host_id, year, sep = "_")),
         genus_species = as_factor(paste(str_sub(host_genus, 1, 1), ". ", host_species, sep = "")),
         gspp = factor(paste0(str_sub(host_genus, 1, 1), str_sub(host_species, 1, 3)),
                       levels = levels(sppord))) %>%
  filter(Abundance > 0.01)# %>% 
  #complete(year, host_id)

pmdf <- pmdf %>%
  mutate(yearlab = case_when(host_id == "12" ~ year,
                             TRUE ~ ""))
  
g2 <- ggplot(pmdf, aes(x = year, y = Abundance)) +
    geom_col(width = 1, aes(fill = its2_type_profile_trunc)) +
    facet_nested(~ gspp + host_site + host_id) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) +
    ylab("Profiles") +
    scale_fill_manual(values = ppal[pmdf$its2_type_profile_trunc], limits = force, drop = TRUE) +
    theme(legend.position = "bottom", legend.key.size = unit(2, "mm"), legend.title = element_blank(),
          legend.margin = margin(1, 0, 0, 0), legend.box.margin = margin(-10, -10, 0, -10),
          legend.key = element_rect(fill = NA),
          plot.margin = margin(0, 5.5, 5.5, 5.5),
          text = element_text(size = 6), legend.text = element_text(size = 4.4),
          panel.background = element_blank(), strip.text.x = element_blank(),
          axis.title.x = element_blank(), axis.text.y = element_blank(), 
          axis.text.x = element_text(angle = 90, vjust = 0.5, size = 3),
          axis.ticks.x = element_line(size = 0.25),
          axis.ticks.y = element_blank(),
          strip.background = element_blank(),
          panel.spacing = unit(0.25, "mm")) +
    guides(fill = guide_legend(ncol = 8))
g2

fig1 <- plot_grid(g1, g2, nrow = 2, label_fontface = "plain", rel_heights = c(0.6, 0.4))

fig1

ggsave(fig1, filename = "Output/Figure1.png", width = 183, height = 70, units = "mm")
```



```{r, }
# which profiles in which species?
favi <- pmdf %>%
  filter(host_family == "Faviidae") %>%
  select(Sample, gspp, Abundance, clade, its2_type_profile) %>%
  group_by(gspp, clade, its2_type_profile) %>%
  summarize(meanAbund = mean(Abundance), n = n()) %>%
  ungroup()
# 
favi %>%
  filter(clade == "B") %>%
  arrange(its2_type_profile) %>%
  arrange(-n) %>%
  print(n = nrow(.))
# 
# vmdf %>% filter(Abundance > 0) %>% filter(DIV == "B14b")
# 
# # Dlab's Breviolum is also found at in one Mcav sample...
# pmdf %>% filter(its2_type_profile == "B1-B1al-B1aj-B1ak-B1aa")
# # Dlab's other Breviolum is also found at in one Mcav sample...
# pmdf %>% filter(its2_type_profile == "B1-B1z-B1aa-B1ad-B1ac-B1ab")
# # Pstr's Breviolum is also found in Mcav, Ofav, Ofra, Ssid
# pmdf %>% filter(its2_type_profile == "B1-B1x-B1t-B14b-B1af-B1ae-B1ag")
# # Cnat's Breviolum is also found in Mcav and Ofra
# pmdf %>% filter(its2_type_profile == "B19/B5af-B19e-B19c-B5ag-B19d-B19f-B40a-B19h")
```


```{r, dom_bk_etc}
# Changes in dominant genus
genus_change <- pmdf %>%
  filter(Abundance >= 0.5) %>%
  group_by(gspp, host_id) %>%
  summarize(n_dom_clades = n_distinct(host_id, clade)) %>%
  filter(n_dom_clades > 1)

# Changes in dominant profile
prof_change <- pmdf %>%
  filter(Abundance >= 0.5) %>%
  group_by(gspp, clade, host_id) %>%
  summarize(n_dom_profs = n_distinct(host_id, clade, its2_type_profile)) %>%
  filter(n_dom_profs > 1)
prof_change %>% distinct(gspp)
# 13 colonies changed dominant profile within genus
# how many total colonies sampled more than one year?
pmdf %>%
  distinct(host_id, year) %>%
  count(host_id) %>%
  count(n > 1)
# 56 colonies sampled more than once, 4 shuffled genera, and 13 shuffled profiles, so 39 did not

####### new organization of all of this

# get abundance of each clade in each sample
clades <- vmdf %>%
  # Get clade level relative abundances
  group_by(gspp, host_site, host_id, year, clade) %>%
  summarize(Abundance = sum(Abundance)) %>%
  group_by(gspp, host_site, host_id, year) %>%
  pivot_wider(names_from = clade, values_from = Abundance) %>%
  ungroup()

# get number of clades over threshold abundance in each sample
nclades <- clades %>%
  nest(data = c(A, B, C, D)) %>%
  mutate(nclades.0.1 = map_dbl(data, ~ sum(. > 0.001)),
         nclades.1   = map_dbl(data, ~ sum(. > 0.01)),
         nclades.10  = map_dbl(data, ~ sum(. > 0.1)))

# how many colonies ever had multiple clades (all time points)?
mult <- nclades %>%
  group_by(gspp, host_site, host_id) %>%
  summarize(mult.0.1 = any(nclades.0.1 > 1),
            mult.1   = any(nclades.1 > 1),
            mult.10  = any(nclades.10 > 1))
table(mult$mult.0.1) # 64/70 (91%) colonies had multiple clades > 0.1% abund
table(mult$mult.1) # 33/70 (47%) colonies had multiple clades > 1% abund
table(mult$mult.10) # 11/70 (16%) colonies had multiple clades > 10% abund

# # did presence of multiple clades vary by coral species?
# mod <- glmer(mult.1 ~ gspp + (1|host_id) + (1|host_site), data = mult, family = "binomial")
# car::Anova(mod)
# anova(mod)
# plot(emmeans(mod, specs = "gspp", type = "response"))
# 
# 
# # how many colonies had a change in the number of clades over time?
# nnclades <- nclades %>%
#   group_by(gspp, host_site, host_id) %>%
#   filter(n() > 1) %>%    # only include colonies sampled more than once over time
#   summarize(nnclades.0.1 = n_distinct(nclades.0.1),
#             nnclades.1   = n_distinct(nclades.1),
#             nnclades.10  = n_distinct(nclades.10))
# table(nnclades$nnclades.0.1 > 1) # 32/56 (57%) colonies had a change in the number of clades > 0.1% abund
# table(nnclades$nnclades.1 > 1) # 27/56 (48%) colonies had a change in the number of clades > 1% abund
# nnclades %>% filter(nnclades.1 > 1) %>% ungroup() %>% count(gspp)
# table(nnclades$nnclades.10 > 1) # 8/56 (14%) colonies had a change in the number of clades > 10% abund
# 
# # did change in number of clades vary by coral species?   NO EFFECT, NOT MUCH STATISTICAL POWER
# df <- nnclades %>%
#   mutate(nnclades.0.1 = nnclades.0.1 > 1)
# mod <- glmer(nnclades.0.1 ~ gspp + (1|host_id) + (1|host_site), data = df, family = "binomial")
# car::Anova(mod)
# anova(mod)
# plot(emmeans(mod, specs = "gspp", type = "response"))
# 
# ggplot(nclades, aes(x = year, y = nclades.0.1)) +
#   geom_jitter(width = 0.1, height = 0.1, alpha = 0.5) +
#   facet_wrap(~ gspp)
# 
# # Model number of clades over time -- filter out Pfur since only one colony sampled over time
# df <- filter(nclades, gspp != "Pfur") %>%
#   mutate(year = factor(year))
# mod <- lmerTest::lmer(nclades.0.1 ~ year * gspp + (1|host_site) + (1|host_id), data = df)
# hist(residuals(mod))
# car::Anova(mod)
# emm <- emmeans(mod, c("year", "gspp"))
# 
# # some evidence for decreasing number of clades per colony over time?
# as_tibble(emm) %>%
#   ggplot(aes(x = year, y = emmean, color = gspp, group = gspp)) +
#   geom_line()
# 
# contrast(emmeans(mod, "year"), "pairwise")
# 
# 
# # which background clades are fluctuating?
# fluc <- clades %>%
#   pivot_longer(cols = c(A, B, C, D), names_to = "clade", values_to = "Abundance") %>%
#   filter(host_id %in% nnclades$host_id[nnclades$nnclades.1 > 1],     # select colonies with changing # clades > 1%
#          Abundance > 0.01) %>%   
#   nest(data = c(year, clade, Abundance))
# 
# maxmin <- fluc %>%
#   mutate(cladecount = map(data, ~ count(., clade)),
#          maxclade = map_chr(cladecount, ~ .$clade[which.max(.$n)]),
#          minclade = map_chr(cladecount, ~ .$clade[which.min(.$n)]))
# maxmin %>%
#   count(maxclade, minclade) %>%
#   arrange(maxclade, minclade)
# 
# # did presence of multiple clades vary by dominant clade?
# test <- clades %>%
#   nest(data = c(year, A, B, C, D)) %>%
#   mutate(sums = map(data, ~ colSums(.[2:5])),
#          dom = map_chr(sums, ~ names(which.max(.))),
#          bk = map(sums, ~ setdiff(names(which(. > 0.01)), names(which.max(.)))))
# 
# test2 <- test %>%
#   select(gspp, host_site, host_id, dom, bk) %>%
#   unnest(bk) %>%
#   right_join(select(test, gspp, host_site, host_id, dom)) %>%
#   mutate(bk = case_when(is.na(bk) ~ "none", TRUE ~ bk)) %>%
#   pivot_wider(names_from = "bk", values_from = "bk") %>%
#   mutate(across(c("B", "D", "C", "A", "none"), ~!is.na(.)))
# 
# 
# mod <- glmer(D ~ dom + (1|gspp) + (1|host_site), data = filter(test2, dom != "D"), family = "binomial")
# anova(mod)
# plot(emmeans(mod, specs = "dom", type = "response"))
# 
# test3 <- test %>%
#   select(gspp, host_site, host_id, dom, bk) %>%
#   unnest(bk) %>%
#   right_join(select(test, gspp, host_site, host_id, dom)) %>%
#   mutate(bk = case_when(is.na(bk) ~ "none", TRUE ~ bk))
# 
# dombk <- test3 %>%
#   group_by(dom) %>%
#   count(bk) %>%
#   pivot_wider(names_from = bk, values_from = n, values_fill = 0) %>%
#   ungroup()
# hi<-dombk %>%
#   select(-dom) %>%
#   as.matrix()
# chisq.test(hi)
# dombk %>% select(dom, A, B, C, D, none)
# 
# 
# 
# gsppbk <- test3 %>%
#   group_by(gspp) %>%
#   count(bk) %>%
#   pivot_wider(names_from = bk, values_from = n)
```

```{r nmds_Fig2}
samples_clades <- vmdf %>%
  # Get clade level relative abundances
  group_by(Sample, gspp, host_site, host_id, year, clade) %>%
  summarize(Abundance = sum(Abundance)) %>%
  group_by(Sample, gspp, host_site, host_id, year) %>%
  ungroup()
samples_with_5pA <- samples_clades %>% filter(clade == "A", Abundance >= 0.05)
samples_with_5pB <- samples_clades %>% filter(clade == "B", Abundance >= 0.05)
samples_with_5pC <- samples_clades %>% filter(clade == "C", Abundance >= 0.05)
samples_with_5pD <- samples_clades %>% filter(clade == "D", Abundance >= 0.05)
#NMDS
# Look at just Symbiodinium
varA <- subset_samples(variants, sample_name %in% samples_with_5pA$Sample)
varA <- subset_taxa(varA, clade == "A")
# Run NMDS for visualization
set.seed(54)
varA.nmds <- ordinate(varA, "NMDS", "bray")
varA.nmds.s <- plot_ordination(varA, varA.nmds, type="samples")
# Look at just Breviolum
varB <- subset_samples(variants, sample_name %in% samples_with_5pB$Sample)
varB <- subset_taxa(varB, clade == "B")
# Run NMDS for visualization
set.seed(54)
varB.nmds <- ordinate(varB, "NMDS", "bray")
varB.nmds.s <- plot_ordination(varB, varB.nmds, type="samples")
# Look at just Cladocopium
varC <- subset_samples(variants, sample_name %in% samples_with_5pC$Sample)
varC <- subset_taxa(varC, clade == "C")
# Run NMDS for visualization
set.seed(55)
varC.nmds <- ordinate(varC, "NMDS", "bray")
varC.nmds.s <- plot_ordination(varC, varC.nmds, type="samples")
# Look at just Durusdinium corals
varD <- subset_samples(variants, sample_name %in% samples_with_5pD$Sample)
varD <- subset_taxa(varD, clade == "D")
# Run NMDS for visualization
set.seed(66)
varD.nmds <- ordinate(varD, "NMDS", "bray")
varD.nmds.s <- plot_ordination(varD, varD.nmds, type="samples")

# Plot NMDS for each genus
clade_nmds <- bind_rows(
  Symbiodinium = varA.nmds.s$data,
  Breviolum = varB.nmds.s$data, 
  Cladocopium = varC.nmds.s$data, 
  Durusdinium = varD.nmds.s$data,
  .id = "sym_genus"
) %>%
  arrange(host_id, year) %>%
  mutate(sym_genus = factor(sym_genus, levels = c("Symbiodinium", "Breviolum", 
                                                  "Cladocopium", "Durusdinium"))) %>%
  ggplot(aes(x = NMDS1, y = NMDS2, color = host_species, shape = host_species)) +
  facet_wrap(~ sym_genus, scales = "free", nrow = 1) +
  geom_point(alpha = 1, size = 2) +
  geom_path(aes(group = host_id), linewidth = 0.2) +
  scale_shape_manual(values = 8:0) +
  theme_custom() +
  theme(legend.position = "bottom",
        strip.background = element_blank(),
        strip.text = element_text(face = "italic"),
        legend.key.size = unit(2, "mm"), legend.title = element_blank(),
          legend.margin = margin(1, 0, 0, 0), legend.box.margin = margin(-10, -10, 0, -10),
          legend.key = element_rect(fill = NA),
          text = element_text(size = 6), legend.text = element_text(size = 4)) +
  guides(shape = guide_legend(ncol = 9))

clade_nmds
ggsave(clade_nmds, filename = "Output/Figure2.png", width = 183, height = 60, units = "mm")

# # bigfig <- plot_grid(fig1, clade_nmds, nrow = 2, labels = "AUTO", rel_heights = c(0.55, 0.45))
# # ggsave(bigfig, filename = "Output/bigfig.png", width = 183, height = 120, units = "mm")
# 
# # LOOK AT ALL CORALS?
# # Run NMDS for visualization
# set.seed(55)
# var.nmds <- ordinate(variants, "NMDS", "bray")
# var.nmds.s <- plot_ordination(variants, var.nmds, type="samples")
# # Plot NMDS
# allnmds <- as_tibble(var.nmds.s$data) %>%
#   arrange(host_id, year) %>%
#   ggplot(aes(x = NMDS1, y = NMDS2, shape = host_species, color = host_species)) +
#   geom_point(alpha = 0.5, size = 2) +
#   geom_path(aes(group = host_id), arrow = arrow(ends = "last", length = unit(1, "mm"))) +
#   scale_shape_manual(values = 1:9) +
#   theme_custom()
# 
# hull_data <- as_tibble(var.nmds.s$data) %>%
#   group_by(host_species) %>% 
#   slice(chull(NMDS1, NMDS2))
# as_tibble(var.nmds.s$data) %>%
#   arrange(host_id, year) %>%
#   ggplot(aes(x = NMDS1, y = NMDS2, shape = host_species, 
#              color = host_species, fill = host_species)) +
#   geom_point(alpha = 0.5, size = 2) +
#   scale_shape_manual(values = 1:9) +
#   theme_custom() +
#   geom_polygon(aes(group = host_species, fill = host_species),
#                data = hull_data, alpha = 0.3, linewidth = 0)# +
#   #stat_density_2d(geom = "polygon", alpha = 0.05, lwd = 0, binwidth = 0.5)
# 
# set.seed(55)
# as_tibble(var.nmds.s$data) %>%
#   arrange(host_id, year) %>%
#   ggplot(aes(x = NMDS1, y = NMDS2)) +
#   geom_point(aes(shape = year, color = year)) +
#   geom_path(aes(group = host_id), alpha = 0.5, arrow = arrow(ends = "last", length = unit(2, "mm"))) +
#   facet_wrap(~host_species, scales = "fixed") +
#   theme_custom()
#   # coord_fixed(xlim = c(-1, 1), ylim = c(-1, 1))
# 
# as_tibble(var.nmds.s$data) %>%
#   group_by(host_id) %>%
#   filter(any(year == "2017")) %>%
#   mutate(NMDS1adj = NMDS1 - NMDS1[year == "2017"],
#          NMDS2adj = NMDS2 - NMDS2[year == "2017"]) %>%
#   ungroup() %>%
#   arrange(host_id, year) %>%
#   ggplot(aes(x = NMDS1adj, y = NMDS2adj)) +
#   geom_point(aes(shape = year, color = year)) +
#   geom_path(aes(group = host_id), alpha = 0.5, arrow = arrow(ends = "last", length = unit(2, "mm"))) +
#   facet_wrap(~host_species) +
#   scale_x_continuous(limits = c(-0.7, 0.7)) +
#   scale_y_continuous(limits = c(-0.7, 0.7)) +
#   theme_custom()



# as_tibble(var.nmds.s$data) %>%
#   arrange(host_id, year) %>%
#   ggplot(aes(x = NMDS1, y = NMDS2, shape = host_species, 
#              color = host_species, fill = host_species)) +
#   geom_point(alpha = 0.5, size = 1) +
#   scale_shape_manual(values = 1:9) +
#   theme_custom() +
#   stat_centroid(aes(xend = NMDS1, yend = NMDS2, group = host_species), 
#                 geom = "segment", crop_other = FALSE, linewidth = 0.25)




as_tibble(varB.nmds.s$data) %>%
  filter(host_species == "siderea")

pmdf2 <- psmelt(p) 
pmdf2 %>% filter(host_id == "51", Abundance > 0.5)
pmdf2 %>% filter(its2_type_profile == "B1-B1z-B1aa-B1ad-B1ac-B1ab", Abundance > 0.5)
pmdf2 %>% filter(its2_type_profile == "B1-B1x-B1t-B14b-B1af-B1ae-B1ag", Abundance > 0.5)
pmdf2 %>% filter(host_species == "cavernosa", Abundance > 0.9)
pmdf2 %>% filter(grepl("C40g", its2_type_profile), Abundance > 0) %>% 
  as_tibble() %>% select(Sample, Abundance, host_id, year, host_species) %>% print(n = nrow(.))
pmdf2 %>% filter(grepl("C3an", its2_type_profile), Abundance > 0) %>% 
  as_tibble() %>% select(Sample, Abundance, host_id, year, host_species,
                         its2_type_profile) %>% print(n = nrow(.))
pmdf2 %>% filter(grepl("C7", its2_type_profile), Abundance > 0.01) %>% 
  as_tibble() %>% select(Sample, Abundance, host_id, year, host_species,
                         its2_type_profile) %>% print(n = nrow(.))
pmdf2 %>% filter(host_species == "strigosa", clade == "C", Abundance > 0.5)

pmdf2 %>% filter(grepl("C3fe", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(grepl("C54a", its2_type_profile), Abundance > 0)

pmdf2 %>% filter(host_species == "siderea", clade == "C", Abundance > 0.5)
pmdf2 %>% filter(grepl("C3a-", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(grepl("C16a", its2_type_profile), Abundance > 0)

pmdf2 %>% filter(host_species == "furcata", clade == "C", Abundance > 0)
pmdf2 %>% filter(grepl("C47", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(grepl("C42", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(grepl("C45a", its2_type_profile), Abundance > 0)


pmdf2 %>% filter(host_species == "natans", clade == "B", Abundance > 0)
pmdf2 %>% filter(grepl("B5af", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(host_species == "strigosa", clade == "B", Abundance > 0)
pmdf2 %>% filter(grepl("B1x", its2_type_profile), Abundance > 0)
```

# Stats to compare species and sites
```{r}
## PERMANOVAs...
varr <- transform_sample_counts(variants, function(x) sqrt(x/sum(x)))  


cnat <- subset_samples(varr, host_species == "natans")
adonis(distance(cnat, "bray") ~ host_site, data = as_tibble(sample_data(cnat)))
plot_bar(cnat, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

dlab <- subset_samples(varr, host_species == "labrynthiformis")
adonis(distance(dlab, "bray") ~ host_site, data = as_tibble(sample_data(dlab)))
plot_bar(dlab, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

pstr <- subset_samples(varr, host_species == "strigosa")
adonis(distance(pstr, "bray") ~ host_site, data = as_tibble(sample_data(pstr)))
plot_bar(pstr, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

mcav <- subset_samples(varr, host_species == "cavernosa")
adonis(distance(mcav, "bray") ~ host_site, data = as_tibble(sample_data(mcav)))
plot_bar(mcav, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

ssid <- subset_samples(varr, host_species == "siderea")
adonis(distance(ssid, "bray") ~ host_site, data = as_tibble(sample_data(ssid)))
plot_bar(ssid, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

ofav <- subset_samples(varr, host_species == "faveolata")
adonis(distance(ofav, "bray") ~ host_site, data = as_tibble(sample_data(ofav)))
plot_bar(ofav, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

oann <- subset_samples(varr, host_species == "annularis")
adonis(distance(oann, "bray") ~ host_site, data = as_tibble(sample_data(oann)))
plot_bar(oann, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

ofra <- subset_samples(varr, host_species == "franksi")
adonis(distance(ofra, "bray") ~ host_site, data = as_tibble(sample_data(ofra)))
plot_bar(ofra, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

pfur <- subset_samples(varr, host_species == "furcata")
adonis(distance(pfur, "bray") ~ host_site, data = as_tibble(sample_data(pfur)))
plot_bar(pfur, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")
```


# How dynamic are coral symbiont communities within individual colonies? (change over time)
### Bray-Curtis diffs within colonies over time
```{r}
# How does B-C dissim change from 2017--2018--2019?
#varr <- transform_sample_counts(variants, function(x) x/sum(x))     # Square-root transform

# Bray-Curtis distance matrix
# bd <- as.matrix(distance(varr, "bray"))
# xy <- t(combn(colnames(bd), 2))
# sc <- as_tibble(data.frame(xy, dist=bd[xy])) %>%
#   separate(X1, into = c("site1", "species1", "host_id1", "year1")) %>%
#   separate(X2, into = c("site2", "species2", "host_id2", "year2")) %>%
#   filter(host_id1 == host_id2)

# sc4 <- sc %>%
#   mutate(yearx = if_else(year1 < year2, year1, year2),
#          yeary = if_else(year1 < year2, year2, year1)) %>%
#   select(host_id = host_id1, host_species = species1,
#          year1 = yearx, year2 = yeary, dist) %>%
#   filter(year1 == 2017) %>%
#   mutate(year2 = as_factor(year2))
# 
# mod <- lmer(log(dist) ~ year2 * host_species + (1|host_id), 
#             data = filter(sc4, host_species != "Pfur"))
# hist(residuals(mod))
# shapiro.test(residuals(mod))
# car::Anova(mod)
# emm <- emmeans(mod, specs = c("year2", "host_species"), type = "response")
# contrast(emm, "pairwise", by = "host_species")
# contrast(emm, "pairwise", by = "year2", adjust = "none")
# 
# 
# f2 <- as_tibble(emm) %>%
#   mutate(emmean = response) %>%
#   full_join(tibble(year2 = "2017", distinct(sc4, host_species), emmean = 0)) %>%
#   mutate(host_species = factor(host_species, levels = c("Mcav", "Ssid", "Oann",
#                                                         "Dlab", "Ofra", "Cnat",
#                                                         "Pstr", "Pfur", "Ofav"))) %>%
#   ggplot(aes(x = year2, y = emmean, color = host_species, group = host_species)) + 
#   geom_point(position = position_dodge(width = 0.3)) +
#   geom_errorbar(aes(ymin = emmean - SE, ymax = emmean + SE), 
#                 position = position_dodge(width = 0.3), width = 0.1) +
#   geom_line(position = position_dodge(width = 0.3)) +
#   theme_custom() +
#   labs(x = "", y = "Bray-Curtis dissimilarity") +
#   theme(legend.text = element_text(face = 'italic'),
#         legend.key.size = unit(2, "mm"),
#         legend.position = c(0.2, 0.85),
#         legend.spacing.y = unit(0.05, "mm")) +
#   guides(color = guide_legend(byrow = TRUE)) +
#   scale_color_discrete(labels = c("M. cavernosa", "S. siderea", "O. annularis", "D. labyrinthiformis",
#                                   "O. franksi", "C. natans", "P. strigosa", "P. furcata", "O. faveolata"),
#                        name = "")
# 
# ggsave(f2, filename = "Output/Figure2.png", width = 85, height = 85, units = "mm")

#variants_nopfur <- subset_samples(variants, host_species != "furcata")



varr <- transform_sample_counts(variants, function(x) sqrt(x/sum(x))) 
#set.seed(55)
set.seed(56)
varr.nmds <- ordinate(varr, "NMDS", "bray")
varr.nmds.s <- plot_ordination(varr, varr.nmds, type="samples")

nmds_fig <- as_tibble(varr.nmds.s$data) %>%
  mutate(gspp = factor(paste0(str_sub(host_genus, 1, 1), str_sub(host_species, 1, 3)))) %>%
  arrange(host_id, year) %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(aes(shape = year, color = year), alpha = 0.6) +
  geom_path(aes(group = host_id), alpha = 0.5, arrow = arrow(ends = "last", length = unit(1, "mm")), lwd = 0.25) +
  facet_wrap(~gspp, labeller = global_labeller) +
  scale_color_manual(values = c('#66c2a5','#fc8d62','#8da0cb')) +
  theme_custom() +
  theme(strip.background = element_blank(),
          strip.text = element_text(face = "italic"),
          text = element_text(size = 7), legend.text = element_text(size = 4),
          axis.text.x = element_text(hjust = 0)) +
  theme(legend.position = "none"); nmds_fig

ggsave(nmds_fig, filename = "Output/Figure3.png", width = 85, height = 85, units = "mm")




# include as supplementary?
as_tibble(varr.nmds.s$data) %>%
  group_by(host_id) %>%
  filter(any(year == "2017")) %>%
  mutate(NMDS1adj = NMDS1 - NMDS1[year == "2017"],
         NMDS2adj = NMDS2 - NMDS2[year == "2017"]) %>%
  ungroup() %>%
  arrange(host_id, year) %>%
  ggplot(aes(x = NMDS1adj, y = NMDS2adj)) +
  geom_point(aes(shape = year, color = year)) +
  geom_path(aes(group = host_id), alpha = 0.5, arrow = arrow(ends = "last", length = unit(2, "mm"))) +
  facet_wrap(~host_species) +
  scale_x_continuous(limits = c(-1, 1)) +
  scale_y_continuous(limits = c(-1, 1)) +
  theme_custom()
```

# Compare between-colony and within-colony Bray-Curtis dissimilarities across species

```{r}
varr <- transform_sample_counts(variants, function(x) sqrt(x/sum(x)))    # Square-root transform

bd <- as.matrix(distance(varr, "bray"))
xy <- t(combn(colnames(bd), 2))
bd2 <- as_tibble(data.frame(xy, dist=bd[xy])) %>%
  separate(X1, into = c("site1", "species1", "host_id1", "year1")) %>%
  separate(X2, into = c("site2", "species2", "host_id2", "year2")) %>%
  filter(species1 == species2) %>%    # only interested in within-species differences
  mutate(group = case_when(host_id1 == host_id2 ~ "within", TRUE ~ "between")) %>%
  filter(group == "within" |
         group == "between" & year1 == year2)



#order species by highest median between-colony difference
brayord <- bd2 %>%
  filter(group == "between") %>%
  group_by(species1) %>%
  summarize(medbray = median(dist)) %>%
  arrange(-medbray) %>%
  mutate(gspp = as_factor(species1))
  
source("https://raw.githubusercontent.com/datavizpyr/data/master/half_flat_violinplot.R")
f4 <- bd2 %>% mutate(gspp = factor(species1, levels = levels(brayord$gspp))) %>%
  ggplot(aes(x = group, y = dist)) +
    facet_wrap(~ gspp, nrow = 2, labeller = global_labeller) +
    geom_flat_violin(adjust = 1, position = position_nudge(x = 0.15), scale = "width",
                     lwd = 0, fill = "grey72") +
    geom_boxplot(width = 0.1, outlier.shape = NA, lwd = 0.2,
                 position = position_nudge(x = 0.15)) +
    geom_jitter(width = 0.075, alpha = 0.2, stroke = 0) +
    theme_custom() +
    theme(strip.background = element_blank(),
          strip.text = element_text(face = "italic"),
          text = element_text(size = 7), legend.text = element_text(size = 4),
          axis.ticks.x = element_blank(),
          axis.text.x = element_text(hjust = 0)) +
    labs(y = "Bray-Curtis Dissimilarity", x = "") +
    coord_cartesian(xlim = c(1.3, 2.1)); f4

ggsave(f4, filename = "Output/Figure4.png", width = 102, height = 80, units = "mm")

bd2 %>%
  group_by(species1, group) %>%
  summarize(median_dist = median(dist)) %>%
  pivot_wider(names_from = group, values_from = median_dist) %>%
  mutate(btw_wit = between/within) %>%
  arrange(btw_wit)

# revised figure 3
bd2 %>% mutate(gspp = factor(species1, levels = levels(brayord$gspp))) %>%
  ggplot(aes(x = group, y = dist)) +
    facet_wrap(~ gspp, labeller = global_labeller, ncol = 5) +
    # geom_flat_violin(adjust = 1, position = position_nudge(x = 0.15), scale = "width",
    #                  lwd = 0, fill = "grey72") +
    geom_boxplot(width = 0.1, outlier.shape = NA, lwd = 0.2,
                 position = position_nudge(x = 0.15)) +
    geom_jitter(width = 0.075, alpha = 0.2, stroke = 0) +
    theme_custom() +
    theme(strip.background = element_blank(),
          strip.text = element_text(face = "italic"),
          text = element_text(size = 7), legend.text = element_text(size = 4),
          axis.ticks.x = element_blank(),
          axis.text.x = element_text(hjust = 0)) +
    labs(y = "Bray-Curtis Dissimilarity", x = "") +
    coord_cartesian(xlim = c(1.3, 2.1))
```



# 3. Ecological data from Pete
```{r}
# genus level dataset
ecodat <- readxl::read_xls("Data/STJ_corals_genus.xls") %>%
  clean_names()

ecomd <- ecodat %>% select(1:5)
ecocomp <- ecodat %>% select(9:ncol(ecodat))

ecolong <- ecodat %>%
  pivot_longer(cols = 6:ncol(ecodat), names_to = "tax")

mod <- lmer(log(value+0.001) ~ tax * factor(year) + (1|quadrat/site), data = ecolong)
car::Anova(mod)

as_tibble(emmeans(mod, specs = c("year", "tax"))) %>%
  filter(tax %in% c("orbicella", "diploria", "montastraea", "montastrea_cav", "colpophyllia", "siderastrea")) %>%
  ggplot(aes(x = year, y = emmean, color = tax, group = tax)) +
  geom_point() +
  geom_line()
  
# species level dataset
coralspp <- readxl::read_xls("Data/STJ_corals_species.xls", skip = 2) %>%
  clean_names()

sppmd <- coralspp %>% select(1:5)
sppcomp <- coralspp %>% select(9:ncol(coralspp))

spplong <- coralspp %>%
  pivot_longer(cols = 6:ncol(coralspp), names_to = "tax")

mod <- lmer(log(value+0.001) ~ tax * factor(year) + (1|quadrat/site), data = spplong)
car::Anova(mod)

as_tibble(emmeans(mod, specs = c("year", "tax"))) %>%
  filter(tax %in% c("percent_orbicella_annularis", "percent_orbicella_faveolata",
                    "diploria_labrynthiformes", "percent_orbicealla_franksi",
                    "percent_montastraea_cavernosa", "percent_pseudodiploria_strigosa", 
                    "percent_colpophyllia_natans", "percent_sidreastrea_siderea")) %>%
  ggplot(aes(x = year, y = emmean, color = tax, group = tax)) +
  geom_point() +
  geom_line()
```



# Summary figures for each species

The top row of bars shows all ITS2 sequence variants, while the bottom row of bars shows the associated ITS2 profiles. Legend refers to profiles only, not variants.

### *Colpophyllia natans*
```{r natans_summary, fig.width = 11}
plot_both2(host_species == "natans")
```

*C. natans* colonies at Cabritte Horn and East Tektite were dominated by a few different *Breviolum* profiles, with low levels of *Cladocopium* and *Durusdinium* that fluctuated randomly over time. At White Point, one colony was dominated by *Cladocopium* and one by *Durusdinium*.

```{r}
# NMDS for C. natans
cnat <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "natans"))
cnat <- prune_taxa(taxa_sums(cnat) != 0, cnat)
cnatr = transform_sample_counts(cnat, function(x) x/sum(x))

cnatdistmat <- vegan::vegdist(sqrt(otu_table(cnatr)), method = "bray")

cnatdf <- as.data.frame(cmdscale(cnatdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(cnat)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(cnatdf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open"))

# Subset just the B-dom corals
cnat <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "natans" & 
                                                     host_id %in% c("12", "5", "33", "35", "60", "61")))
cnat <- prune_taxa(taxa_sums(cnat) != 0, cnat)
cnatr = transform_sample_counts(cnat, function(x) x/sum(x))

cnatdistmat <- vegan::vegdist(sqrt(otu_table(cnatr)), method = "bray")

cnatdf <- as.data.frame(cmdscale(cnatdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(cnat)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(cnatdf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open")) +
  labs(title = "Just Breviolum-dominated C. natans")
```

### *Diploria labyrinithiformis*
```{r labrynthiformis_summary, fig.width = 11}
plot_both2(host_species == "labrynthiformis")
```

One colony at East Tektite (57) was dominated by *Cladocopium*, but otherwise *D. labyrinthiformis* was dominated by several distinct *Breviolum* profiles. These mostly remained the same over time within colonies, but changed for a few (colony 36, 37, 62). Will have to dig into these to see if these may be real changes or analytical artifacts. 

```{r}
# NMDS for D. lab
dlab <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "labrynthiformis"))
dlab <- prune_taxa(taxa_sums(dlab) != 0, dlab)
dlabr = transform_sample_counts(dlab, function(x) sqrt(x/sum(x)))

dlabdistmat <- vegan::vegdist(sqrt(otu_table(dlabr)), method = "bray")

dlabdf <- as.data.frame(cmdscale(dlabdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(dlab)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(dlabdf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open"))

# Just B-dom Dlab
dlab <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "labrynthiformis" &
                                                     host_id %in% c("13", "50", "36", "37", "62", "20", "23")))
dlab <- prune_taxa(taxa_sums(dlab) != 0, dlab)
dlabr = transform_sample_counts(dlab, function(x) x/sum(x))

dlabdistmat <- vegan::vegdist(sqrt(otu_table(dlabr)), method = "bray")

dlabdf <- as.data.frame(cmdscale(dlabdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(dlab)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(dlabdf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open")) +
  labs(title = "Just Breviolum-dominated D.lab.")
```

### *Porites furcata*
```{r labrynthiformis_summary, fig.width = 11}
plot_both2(host_species == "furcata")
```

```{r}
# NMDS for P. fur
pfur <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "furcata"))
pfur <- prune_taxa(taxa_sums(pfur) != 0, pfur)
pfurr = transform_sample_counts(pfur, function(x) x/sum(x))

pfurdistmat <- vegan::vegdist(sqrt(otu_table(pfurr)), method = "bray")

pfurdf <- as.data.frame(cmdscale(pfurdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(pfur)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(pfurdf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open"))
```

### *Montastraea cavernosa*
```{r cavernosa_summary, fig.width = 11}
plot_both2(host_species == "cavernosa")
```

*M. cavernosa* is dominated by a couple of different *Cladocopium* profiles with some background *Breviolum*. Two colonies at White Point have a distinct *Cladocopium* profile, but there doesn't appear to be any particular change over time.

```{r}
# NMDS for Mcav
mcav <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "cavernosa"))
mcav <- prune_taxa(taxa_sums(mcav) != 0, mcav)
mcavr = transform_sample_counts(mcav, function(x) sqrt(x/sum(x)))

mcavdistmat <- vegan::vegdist(sqrt(otu_table(mcavr)), method = "bray")

mcavdf <- as.data.frame(cmdscale(mcavdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(mcav)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(mcavdf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open"))
```

### *Orbicella spp.*
```{r orbicella_summary, fig.width = 11}
plot_both2(host_genus == "Orbicella")
```

*Orbicella* colonies are sometimes dominated by *Symbiodinium*, *Breviolum*, and *Cladocopium*. Most stay the same over time, but some colonies (24 and 64, both at White Point), have mixtures of *Symbiodinium* and  *Cladocopium* in 2017, and then only *Cladocopium* in 2019. Another colony at White Point (17) appears to change between two *Cladocopium* profiles. At other sites, little evidence of change over time.

```{r}
# NMDS for Orbicella
orb <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_genus == "Orbicella"))
orb <- prune_taxa(taxa_sums(orb) != 0, orb)
orbr = transform_sample_counts(orb, function(x) x/sum(x))

orbdistmat <- vegan::vegdist(sqrt(otu_table(orbr)), method = "bray")

orbdf <- as.data.frame(cmdscale(orbdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(orb)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, host_species, collection_depth) %>%
  arrange(host_id, year)

ggplot(orbdf, aes(x = V1, y = V2, color = host_species)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open"))


# Just Orb that are Cdom the whole time
orb <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_genus == "Orbicella" & 
                                        host_id %in% c("3", "47", "56", "80", "81", "40", "18", "65", "67")))
orb <- prune_taxa(taxa_sums(orb) != 0, orb)
orbr = transform_sample_counts(orb, function(x) x/sum(x))

orbdistmat <- vegan::vegdist(sqrt(otu_table(orbr)), method = "bray")

orbdf <- as.data.frame(cmdscale(orbdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(orb)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, host_species, collection_depth) %>%
  arrange(host_id, year)

ggplot(orbdf, aes(x = V1, y = V2, color = host_species)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open")) +
  labs(title = "Just Cladocopium-dominated Orbicella")

# Orb that had multiple clades
orb <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_genus == "Orbicella" & 
                                        host_id %in% c("17", "24", "64")))
orb <- prune_taxa(taxa_sums(orb) != 0, orb)
orbr = transform_sample_counts(orb, function(x) x/sum(x))

orbdistmat <- vegan::vegdist(sqrt(otu_table(orbr)), method = "bray")

orbdf <- as.data.frame(cmdscale(orbdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(orb)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, host_species, collection_depth) %>%
  arrange(host_id, year)

ggplot(orbdf, aes(x = V1, y = V2, color = host_species)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open")) +
  labs(title = "Orbicella that changed dominant clade")
```

### *Pseudodiploria strigosa*
```{r strigosa_summary, fig.width = 11}
plot_both2(host_species == "strigosa")
```

*P. strigosa* colonies contain mostly *Breviolum* profiles, with some evidences for changes among *Breviolum* types over time. Two colonies at East Tektite (45 and 54) are dominated by *Cladocopium*.

```{r}
# NMDS for Pstr
pstr <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "strigosa"))
pstr <- prune_taxa(taxa_sums(pstr) != 0, pstr)
pstrr = transform_sample_counts(pstr, function(x) x/sum(x))

pstrdistmat <- vegan::vegdist(sqrt(otu_table(pstrr)), method = "bray")

pstrdf <- as.data.frame(cmdscale(pstrdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(pstr)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(pstrdf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open"))

# Just Bdom Pstr
pstr <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "strigosa" & 
                                host_id %in% c("10", "82", "42", "28", "66", "68")))
pstr <- prune_taxa(taxa_sums(pstr) != 0, pstr)
pstrr = transform_sample_counts(pstr, function(x) x/sum(x))

pstrdistmat <- vegan::vegdist(sqrt(otu_table(pstrr)), method = "bray")

pstrdf <- as.data.frame(cmdscale(pstrdistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(pstr)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(pstrdf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open")) +
  labs(title = "Just Breviolum-dominated P.str.")
```

### *Siderastrea siderea*
```{r siderea_summary, fig.width = 11}
plot_both2(host_species == "siderea")
```

*S. siderea* hosts mainly *Cladocopium* profiles that do not change over time. A couple of colonies (7 and 73) may have fluctuating levels of multiple *Cladocopium* types. One colony at Cabritte Horn (51) is dominated by *Breviolum*.

```{r}
# NMDS for Ssid
ssid <- merge_phyloseq(tax_table(variants), otu_table(variants), 
                       subset(sample_data(variants), host_species == "siderea" & host_id != "RANDOM"))
ssid <- prune_taxa(taxa_sums(ssid) != 0, ssid)
ssidr = transform_sample_counts(ssid, function(x) sqrt(x/sum(x)))

ssiddistmat <- vegan::vegdist(sqrt(otu_table(ssidr)), method = "bray")

ssiddf <- as.data.frame(cmdscale(ssiddistmat, k = 2)) %>%
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(ssid)) %>%
  as_tibble() %>%
  select(V1, V2, sample_name, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

ggplot(ssiddf, aes(x = V1, y = V2, color = host_id)) +
  geom_point(aes(shape = year)) +
  geom_path(aes(group = host_id), arrow = arrow(length = unit(0.1, "inches"), type = "open"))
```

# Detailed figures for each coral colony

### Colpophyllia natans
```{r}
cnat <- data.frame(sam) %>%
  distinct(host_id, host_genus, host_species, host_site) %>%
  filter(host_species == "natans")

for (i in 1:nrow(cnat)) {
  p <- plot_both(mycol = cnat[i, "host_id"])
  print(p)
}
```

### Diploria labyrinthiformis
```{r}
dlab <- data.frame(sam) %>%
  distinct(host_id, host_genus, host_species, host_site) %>%
  filter(host_species == "labrynthiformis")

for (i in 1:nrow(dlab)) {
  p <- plot_both(mycol = dlab[i, "host_id"])
  print(p)
}
```

### Montastraea cavernosa
```{r}
mcav <- data.frame(sam) %>%
  distinct(host_id, host_genus, host_species, host_site) %>%
  filter(host_species == "cavernosa")

for (i in 1:nrow(mcav)) {
  p <- plot_both(mycol = mcav[i, "host_id"])
  print(p)
}
```


### Orbicella spp.
```{r}
orb <- data.frame(sam) %>%
  distinct(host_id, host_genus, host_species, host_site) %>%
  filter(host_genus == "Orbicella")

for (i in 1:nrow(orb)) {
  p <- plot_both(mycol = orb[i, "host_id"])
  print(p)
}
```

### Pseudodiploria strigosa
```{r}
pstr <- data.frame(sam) %>%
  distinct(host_id, host_genus, host_species, host_site) %>%
  filter(host_species == "strigosa")

for (i in 1:nrow(pstr)) {
  p <- plot_both(mycol = pstr[i, "host_id"])
  print(p)
}
```

### Siderastrea siderea
```{r}
ssid <- data.frame(sam) %>%
  distinct(host_id, host_genus, host_species, host_site) %>%
  filter(host_species == "siderea")

for (i in 1:nrow(ssid)) {
  p <- plot_both(mycol = ssid[i, "host_id"])
  print(p)
}
```



# 1. Symbiont community variability among and within coral species - beta diversity

### Ordination for all coral samples
```{r, include = F, eval = F}
# Convert to relative abundance - and square root transform
varr <- transform_sample_counts(variants, function(x) sqrt(x/sum(x)))     # Square-root transform

# Run NMDS for visualization
set.seed(53)
varr.nmds <- ordinate(varr, "NMDS", "bray")
varr.nmds.s <- plot_ordination(varr, varr.nmds, type="samples")
varr.nmds.t <- plot_ordination(varr, varr.nmds, type="taxa")

# Calculate centroids for each species for plotting
centroids <- varr.nmds.s$data %>%
  group_by(host_species) %>%
  summarise(across(starts_with("NMDS"), mean, .names = "o{.col}")) %>%
  right_join(varr.nmds.s$data)

# Plot NMDS
## Version with only coral samples and centroids
nmds_plot <- ggplot(centroids, aes(x = NMDS1, y = NMDS2)) +
  #geom_point(data = varr.nmds.t$data, aes(fill = clade, shape = clade), stroke = NA, alpha = 0.5) +
  geom_point() +
  geom_segment(aes(xend = oNMDS1, yend = oNMDS2, color = host_species)) +
  theme_custom() +
  scale_shape_manual(values = 21:25) +
  scale_color_corals('color') +
  scale_color_corals('fill'); nmds_plot

## Version with coral samples and centroids overlaying symbiont clade contours
nmds_plot <- as_tibble(varr.nmds.t$data) %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  stat_density_2d(aes(fill = clade),
                  data = filter(as_tibble(varr.nmds.t$data), clade != "G"),
                  geom = "polygon", bins = 10, alpha = 0.1, lwd = 0) +
  theme_custom() +
  # # Add coral species ordispiders
  geom_point(data = centroids, aes(color = host_species)) +
  geom_segment(data = centroids, aes(xend = oNMDS1, yend = oNMDS2, color = host_species), alpha = 0.8) +
  scale_fill_manual(values = c('#ffff99', '#b2df8a', '#a6cee3', '#fb9a99')) +
  #scale_color_manual(values = c('#1f78b4', '#33a02c', '#e31a1c', '#b15928', '#ff7f00', '#cab2d6', '#6a3d9a', '#fdbf6f'))
  scale_color_corals('color'); nmds_plot

distmat <- vegan::vegdist(otu_table(varr), "bray")
# Calculate beta diversity as distance to centroid (PCoA) for each host species
bd <- vegan::betadisper(d = distmat, group = sample_data(varr)$host_species,
                        type = "centroid", bias.adjust = FALSE)
permutest(bd, pairwise = TRUE)
bdres <- as_tibble(sample_data(varr)) %>%
  select(host_id, year, sample_name, host_species) %>%
  mutate(dist_to_centroid = bd$distances)

# Analyze differences in distance to centroid among coral species
mod <- lm(dist_to_centroid ~ host_species, data = bdres)
anova(mod)
### Are these stats valid, or do we need stats from betadisper() / permutation tests?
emm <- emmeans(mod, specs = "host_species") %>%
  as_tibble() %>%
  mutate(host_species = fct_reorder(host_species, emmean, .desc = TRUE))

bdres %>%
  mutate(host_species = factor(host_species, levels = levels(emm$host_species))) %>%
  ggplot(aes(x = host_species, y = dist_to_centroid, color = host_species)) +
    geom_point() +
    geom_point(data = emm, aes(y = emmean), pch = 5, size = 5) +
    geom_errorbar(data = emm, aes(y = emmean, ymin = emmean - SE, ymax = emmean + SE), width = 0.25) +
    theme_custom() +
    scale_color_corals('color')

# Plot beta diversity
bd_plot <- ggplot(emm, aes(x = host_species, y = emmean, color = host_species)) +
  geom_col(aes(fill = host_species), width = 0.8) +
  geom_point() +
  geom_errorbar(aes(ymin = emmean - SE, ymax = emmean + SE), width = 0) +
  geom_text(aes(label = host_species, y = 0.01), color = "black", hjust = 0, angle = 90, size = 3) +
  theme_custom() +
  scale_color_corals('color') +
  scale_color_corals('fill') +
  theme(axis.text.x = element_text(color = "white")) +
  theme(legend.position = "none") +
  labs(x = "Coral species", y = "Distance to centroid"); bd_plot

# Multipanel figure
legend <- get_legend(nmds_plot)
nmds_plot_nolegend <- nmds_plot + theme(legend.position = "none")
right_side <- plot_grid(legend, bd_plot, ncol = 1)
cowplot::plot_grid(nmds_plot_nolegend, right_side, nrow = 1, rel_widths = c(0.7, 0.3))
```



# 2b. Which symbionts are changing over time within colonies/species?
```{r mcmc.otu}
library(MCMC.OTU)


# MCAV
physeq <- subset_samples(variants, host_species == "cavernosa")
otus <- data.frame(otu_table(physeq), check.names=F)
mcmc <- data.frame(year=sample_data(physeq)$year, host_species=sample_data(physeq)$host_species, 
                   host_id = sample_data(physeq)$host_id, sample=rownames(sample_data(physeq)),
                   otus, check.names=F)
goods <- purgeOutliers(mcmc, count.columns=5:ncol(mcmc), otu.cut=0.001)    # Remove OTUs comprising < 1% total counts
gss <- otuStack(goods, count.columns=c(5:length(goods[1, ])), condition.columns=c(1:4)) # Stack the data
mm <- mcmc.otu(fixed="year", data=gss, nitt=55000, thin=50, burnin=5000)  # Fit the model 
  
acpass <- otuByAutocorr(mm, gss)  # Selecting the OTUs that were modeled reliably
smm0 <- OTUsummary(mm, gss, otus=acpass, summ.plot=FALSE)   # Diff & p-vals btw all pairs of factor combinations
smmA <- padjustOTU(smm0)    # Adjust p-values for multiple comparisons
sigs <- signifOTU(smmA, p.cutoff=0.1)    # Significant OTUs at FDR level

smm1 <- OTUsummary(mm, gss, otus=sigs, summ.plot=FALSE)
taxinfo <- data.frame(otu=rownames(tax_table(physeq)),
                      clade=data.frame(tax_table(physeq))$clade,
                      DIV=data.frame(tax_table(physeq))$DIV)
summary <- merge(smm1$summary, taxinfo, by="otu")

summary %>%
  mutate(perc = 10^mean * 100)

ggplot(summary, aes(x = year, y = mean)) + geom_point()
# B1 goes up in 2018 then back down in 2019. This is just one seq though... can we glom to clade level to get an idea of what the relative abundance of all B seqs was in Mcav over time?

# SIDEREA - nothing significant
# ANNULARIS / ORBICELLA - nothing significant
# all corals? - nothing significant
physeq <- subset_samples(variants, host_id == 7)
otus <- data.frame(otu_table(physeq), check.names=F)
mcmc <- data.frame(year=sample_data(physeq)$year, host_species=sample_data(physeq)$host_species, 
                   host_id = sample_data(physeq)$host_id, sample=rownames(sample_data(physeq)),
                   otus, check.names=F)
goods <- purgeOutliers(mcmc, count.columns=5:ncol(mcmc), otu.cut=0.001)    # Remove OTUs comprising < 0.1% total counts
gss <- otuStack(goods, count.columns=c(5:length(goods[1, ])), condition.columns=c(1:4)) # Stack the data
mm <- mcmc.otu(fixed="year", data=gss, nitt=55000, thin=50, burnin=5000)  # Fit the model 
  
acpass <- otuByAutocorr(mm, gss)  # Selecting the OTUs that were modeled reliably
smm0 <- OTUsummary(mm, gss, otus=acpass, summ.plot=FALSE)   # Diff & p-vals btw all pairs of factor combinations
smmA <- padjustOTU(smm0)    # Adjust p-values for multiple comparisons
sigs <- signifOTU(smmA, p.cutoff=0.7)    # Significant OTUs at FDR level

smm1 <- OTUsummary(mm, gss, otus=sigs, summ.plot=FALSE)
taxinfo <- data.frame(otu=rownames(tax_table(physeq)),
                      clade=data.frame(tax_table(physeq))$clade,
                      DIV=data.frame(tax_table(physeq))$DIV)
summary <- merge(smm1$summary, taxinfo, by="otu")

summary %>%
  mutate(perc = 10^mean * 100)

ggplot(summary, aes(x = year, y = mean)) + geom_point()
```

```{r mcmc.otu2}
# maybe need to pivot even wider with all OTU_Year combos as columns?
widedat <- as_tibble(select(data.frame(sample_data(variants)), 
                            host_id, year, sample_name, host_genus, host_species)) %>%
  full_join(as_tibble(data.frame(otu_table(varr)), rownames = "sample_name")) %>%
  select(-sample_name) %>%
  group_by(host_id) %>%
  filter(n() == 3) %>%
  ungroup() %>%
  pivot_wider(names_from = year, values_from = 5:ncol(.))
widedistmat <- vegdist(widedat[,4:ncol(widedat)], method = "manhattan", na.rm = TRUE)

clust <- hclust(widedistmat)
plot(clust, labels = paste0(widedat$host_species, widedat$host_id))
grp <- cutree(clust, k = 5)
plot(clust, labels = grp)

#### OK cool, now try MCMC.OTU on groups of corals that hclust identified as similar over time
#natans33,60,12,5 --> nothing sig
#siderea7/46,41,44 --> nothing sig
# cavernosa49,6,39 --> B1 sig
#natans52+strigosa45 --> nothing sig
#orbicella80,81,40,3 --> nothing sig
physeq <- subset_samples(variants, host_species == "cavernosa")
otus <- data.frame(otu_table(physeq), check.names=F)
mcmc <- data.frame(year=sample_data(physeq)$year, host_species=sample_data(physeq)$host_species, 
                   host_id = sample_data(physeq)$host_id, sample=rownames(sample_data(physeq)),
                   otus, check.names=F)
goods <- purgeOutliers(mcmc, count.columns=5:ncol(mcmc), otu.cut=0.001)    # Remove OTUs comprising < 0.1% total counts
gss <- otuStack(goods, count.columns=c(5:length(goods[1, ])), condition.columns=c(1:4)) # Stack the data
mm <- mcmc.otu(fixed="year", data=gss, nitt=55000, thin=50, burnin=5000)  # Fit the model 
  
acpass <- otuByAutocorr(mm, gss)  # Selecting the OTUs that were modeled reliably
smm0 <- OTUsummary(mm, gss, otus=acpass, summ.plot=FALSE)   # Diff & p-vals btw all pairs of factor combinations
smmA <- padjustOTU(smm0)    # Adjust p-values for multiple comparisons
sigs <- signifOTU(smmA, p.cutoff=.1)    # Significant OTUs at FDR level

smm1 <- OTUsummary(mm, gss, otus=sigs, summ.plot=FALSE)
taxinfo <- data.frame(otu=rownames(tax_table(physeq)),
                      clade=data.frame(tax_table(physeq))$clade,
                      DIV=data.frame(tax_table(physeq))$DIV)
summary <- merge(smm1$summary, taxinfo, by="otu")

summary %>% mutate(perc = 10^mean * 100)
#MCMC.OTU identifies B1 and B19 as increasing in 2018, then decreasing again in 2019
```

```{r deseq2}
library(DESeq2)

# Use DESeq2 to test differences between years for different sets of colonies
out <- deseqDA(host_ids = c(52, 45))#natans52+strigosa45 --> nothing sig
out <- deseqDA(host_ids = c(33,60,12,5)) #natans33,60,12,5 --> nothing compelling
out <- deseqDA(host_ids = c(46, 41, 44)) #siderea7/46,41,44 --> nothing compelling
out <- deseqDA(host_ids = c(49,6,39))# cavernosa49,6,39 --> Bs and Cs sig
out <- deseqDA(host_ids = c(80, 81, 40, 3)) #orbicella80,81,4,3 --> nothing compelling
out <- deseqDA(host_ids = c(49, 58, 6, 34, 39, 55, 19, 25, 69, 70))   # ALL CAVERNOSA B1/B19
out
ggplot(out, aes(x = year, y = count, color = host_id, group = host_id)) +
  facet_wrap(~variant) +
  geom_point()# +
  #geom_line()

#DESeq2 also identifies B1 and B19 in Mcavernosa going up then down again. Also a couple C seqs do same.
```



# PCoA version
```{r}
varr <- transform_sample_counts(variants, function(x) sqrt(x/sum(x)))     # Square-root transform

# Bray-Curtis distance matrix
distmat <- vegan::vegdist(otu_table(varr), method = "bray")

# Principal Coordinates Analysis
varrdf <- as.data.frame(cmdscale(distmat, k = nsamples(varr) - 1)) %>%      # Specify k dimensions   #####
  mutate(sample_name = rownames(.)) %>%
  left_join(sample_data(varr)) %>%
  as_tibble() %>%
  select(starts_with("V"), host_species, host_id, year, host_site, collection_depth) %>%
  arrange(host_id, year)

# Calculate distance moved (in ordination space) from year to year for each colony 
dists <- varrdf %>%
  group_by(host_id) %>%
  #filter(n() == 3) %>%
  nest(data = c(year, starts_with("V"))) %>%
  mutate(years = map(data, ~.[,1]),
         points = map(data, ~.[,-1])) %>%
  mutate(d = map(points, dist),
         pd = map(d, broom::tidy)) %>%
  unnest(pd) %>%
  mutate(y1 = map2_chr(item1, years, ~case_when(.x == 1 ~ .y[[1]][1],
                                                .x == 2 ~ .y[[1]][2])),
         y2 = map2_chr(item2, years, ~case_when(.x == 2 ~ .y[[1]][2],
                                                .x == 3 ~ .y[[1]][3]))) %>%
  select(host_id, host_species, host_site, y1, y2, distance) %>%
  rowwise() %>%
  mutate(period = paste(y1, y2, sep = "-"))

# Plot movement by species
dists %>%
  mutate(gap = as.numeric(y2) - as.numeric(y1)) %>%
  filter(gap == min(gap)) %>%                         # selects 1-year gaps only for cols sampled 3x
  ggplot(aes(x = host_species, y = distance)) +
  geom_violin() +
  geom_point(aes(color = host_site))

# Plot movement by year / species
ggplot(dists, aes(x = interaction(y1,y2), y = distance, color = host_species)) +
  geom_point()


mod <- lmer(log(distance) ~ period * host_species + (1|host_id), data = filter(dists, period %in% c("2017-2018", "2017-2019")))
hist(residuals(mod))
car::Anova(mod)
emm <- emmeans(mod, specs = c("period", "host_species"), type = "response")
contrast(emm, by = "host_species")

# Get model means and combine with zeros as baseline
ord18 <- as_tibble(emm) %>% filter(period == "2017-2018") %>% arrange(-response)
ord18

res <- as_tibble(emm) %>%
  filter(period %in% c("2017-2018", "2017-2019")) %>%
  full_join(tibble(period = "2017", distinct(varrdf, host_species), response = 0)) %>%
  filter(host_species != "furcata") %>%
  mutate(host_species = factor(host_species, levels = c("cavernosa", "siderea", "annularis", "natans",
                                                        "labrynthiformis", "franksi", "faveolata", "strigosa")))

# Plot
res %>%
  ggplot(aes(x = period, y = response, color = host_species, group = host_species)) + 
  geom_point(position = position_dodge(width = 0.3)) +
  geom_errorbar(aes(ymin = response - SE, ymax = response + SE), position = position_dodge(width = 0.3), width = 0.1) +
  geom_line(position = position_dodge(width = 0.3)) +
  theme_custom()

```