---
title: "St. John symbiont time series"
author: "R. Cunning"
date: "2/16/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r libraries}

library(janitor)
library(phyloseq)  # BiocManager::install("phyloseq")
library(cowplot)
library(ggrepel)
library(scales)
library(RColorBrewer)
library(MASS)
library(lme4)
library(emmeans)
library(tidyverse)
library(vegan)
library(ggh4x)
library(stringi)

# library(conflicted)
# conflict_prefer("select", "dplyr")
# conflict_prefer("filter", "dplyr")
# conflict_prefer("rename", "dplyr")
```

# Load Data
```{r profiles, include = FALSE}
sam0 <- read_csv("Data/STJ_Metadata_2017_2019.csv") %>%
  clean_names()
sam1 <- as.matrix(sam0)
rownames(sam1) <- sam0$sample_name
sam <- sample_data(data.frame(sam1))

tax0 <- read_tsv(
  file = "Data/20200623_lenz/its2_type_profiles/111_20200624_2020-06-24_03-40-50.238182.profiles.absolute.abund_and_meta.txt",
  n_max = 6) %>%
  dplyr::select(-2) %>% 
  gather(UID, value, -1) %>% 
  spread(1, value) %>%
  clean_names()
tax1 <- as.matrix(tax0[, -1], dimnames = list(tax0$uid, colnames(tax0[-1])))
rownames(tax1) <- tax0$uid
tax <- tax_table(tax1)

otu0 <- read_tsv(
  file = "Data/20200623_lenz/its2_type_profiles/111_20200624_2020-06-24_03-40-50.238182.profiles.absolute.abund_and_meta.txt") %>% 
  rename(sample_name = X2) %>%
  select(-1) %>%
  slice(7:n()) %>%
  mutate_at(2:ncol(.), as.numeric)
otu1 <- as.matrix(otu0[, -1])
rownames(otu1) <- otu0$sample_name
otu <- otu_table(otu1, taxa_are_rows = FALSE)

profiles <- phyloseq(otu, tax, sam)
```

```{r variants, include = FALSE}
taxnames <- read_tsv(
  file = "Data/20200623_lenz/post_med_seqs/111_20200624_2020-06-24_03-40-50.238182.seqs.absolute.abund_and_meta.txt",
  n_max = 0) %>%
  select(-(1:33)) %>%
  names(.)
tax0 <- data_frame(
  DIV = taxnames,
  clade = str_extract(DIV, "[A-Z]")
)
tax1 <- as.matrix(tax0)
rownames(tax1) <- tax0$DIV
tax <- tax_table(tax1)

otu0 <- read_tsv(
  file = "Data/20200623_lenz/post_med_seqs/111_20200624_2020-06-24_03-40-50.238182.seqs.absolute.abund_and_meta.txt") %>% 
  select(-1, -(3:33))
otu1 <- as.matrix(otu0[, -1])
rownames(otu1) <- otu0$sample_name
otu <- otu_table(otu1, taxa_are_rows = FALSE)

variants <- phyloseq(otu, tax, sam)

tax_table(variants)
```

```{r filter_save, include = FALSE}
# Filter out RANDOM samples (non-tagged colonies)
variants <- subset_samples(variants, host_id != "RANDOM")
variants <- prune_taxa(taxa_sums(variants) != 0, variants)
profiles <- subset_samples(profiles, host_id != "RANDOM")
profiles <- prune_taxa(taxa_sums(profiles) != 0, profiles)

save(profiles, variants, file = "data/Symb_phyloseq.RData")
```

# Create custom functions, color palettes, and ggplot themes
```{r color_palettes, include = FALSE}
# Create standardized color pallettes for plotting profiles and variants

#ppal <- sample(allcolors, ntaxa(profiles))
#ppal <- hue_pal(l = 65)(ntaxa(profiles))
# qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
# col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
# set.seed(555)
# ppal <- sample(col_vector, ntaxa(profiles))
# ppal <- setNames(ppal, data.frame(tax_table(profiles))$its2_type_profile)
# 
# allcolors <- grDevices::colors()
# set.seed(888)
# vpal <- sample(allcolors, ntaxa(variants))
# #vpal <- hue_pal(l = 65)(ntaxa(variants))
# vpal <- setNames(vpal, taxa_names(variants))
# 
# 

# Coral colors

```

```{r plotting_functions, include = FALSE}
# Plotting functions
plot_both <- function(mycol) {
  s <- subset(sample_data(variants), host_id == mycol)
  v <- merge_phyloseq(tax_table(variants), otu_table(variants), s)
  p <- merge_phyloseq(tax_table(profiles), otu_table(profiles), s)
  v <- prune_taxa(taxa_sums(v) != 0, v)
  p <- prune_taxa(taxa_sums(p) != 0, p)
  vmdf <- psmelt(v)
  seqdepth <- vmdf %>%
    group_by(year) %>%
    summarize(sum = paste0("n=", sum(Abundance)))
  ti <- paste0(vmdf[1, c("host_genus", "host_species", "host_id", "host_site")], collapse = " ")
  v_plot <- ggplot(vmdf, aes(x = year, y = Abundance, fill = DIV, label = DIV)) + 
    geom_bar(stat = "identity", position = "fill", color = "black", lwd = 0.1) + 
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)), labels = scales::percent) +
    scale_fill_manual(values = vpal[vmdf$OTU]) +
    geom_text(aes(size = Abundance), position = position_fill(vjust = 0.5)) +
    geom_text(data = seqdepth, aes(x = year, y = 1.01, label = sum), vjust = 0, size = 3, inherit.aes = FALSE) +
    theme(legend.position = "none", panel.background = element_blank(), 
          axis.title.x = element_blank(), axis.title.y = element_blank()) +
    coord_cartesian(ylim = c(0, 1))
  pmdf <- psmelt(p)
  p_plot <- ggplot(pmdf, aes(x = year, y = Abundance, fill = its2_type_profile, label = its2_type_profile)) +
    geom_bar(stat = "identity", position = "fill", color = "black", lwd = 0.1) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)), labels = scales::percent) +
    scale_fill_manual(values = ppal[pmdf$its2_type_profile]) +
    geom_label_repel(position = position_fill(vjust = 0.5), force = 2,
                     point.size = NA, xlim = c(-0.1, 3.6), ylim = c(0.1,0.9), size = 2) +
    theme(legend.position = "none", panel.background = element_blank(), 
          axis.title.x = element_blank(), axis.title.y = element_blank()) +
    coord_cartesian(clip = "off")
  plot_row <- plot_grid(p_plot, v_plot, labels = c("ITS2 Profiles", "ITS2 Variants"), label_fontface = "plain")
  title <- ggdraw() + 
    draw_label(ti, fontface = 'bold', x = 0, hjust = 0) +
    theme(plot.margin = margin(0, 0, 0, 7))
  plot_grid(title, plot_row, ncol = 1, rel_heights = c(0.1, 1))
}

#plot_both(mycol = all3[1])

# Plot by species and site
# ordf <- data.frame(sam) %>%
#   distinct(host_id, host_genus, host_species, host_site) %>%
#   filter(host_id != "RANDOM") %>%
#   filter(host_id %in% all3) %>%
#   arrange(host_genus, host_species, host_site)



# new plotting function to condense by species rather than by individual
plot_both2 <- function(...) {
  s <- subset(sample_data(variants), ...)
  v <- merge_phyloseq(tax_table(variants), otu_table(variants), s)
  p <- merge_phyloseq(tax_table(profiles), otu_table(profiles), s)
  v <- prune_taxa(taxa_sums(v) != 0, v)
  p <- prune_taxa(taxa_sums(p) != 0, p)
  vmdf <- psmelt(v)
  seqdepth <- vmdf %>%
    group_by(year) %>%
    summarize(sum = paste0("n=", sum(Abundance)))
  ti <- paste0(vmdf[1, c("host_genus", "host_species")], collapse = " ")
  v_plot <- ggplot(vmdf, aes(x = year, y = Abundance, fill = DIV, label = DIV)) + 
    geom_bar(stat = "identity", position = "fill") + 
    facet_wrap(~ host_site + host_id, nrow = 1) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)), labels = scales::percent) +
    ylab("ITS2 sequences") +
    scale_fill_manual(values = vpal[vmdf$OTU]) +
    # geom_text(aes(size = Abundance), position = position_fill(vjust = 0.5)) +
    # geom_text(data = seqdepth, aes(x = year, y = 1.01, label = sum), vjust = 0, size = 3, inherit.aes = FALSE) +
    theme(legend.position = "none", panel.background = element_blank(), 
          axis.title.x = element_blank(), axis.text.x = element_blank(),
          axis.text.y = element_blank(), axis.ticks.y = element_blank()) +
    coord_cartesian(ylim = c(0, 1))
  pmdf <- psmelt(p)
  p_plot <- ggplot(pmdf, aes(x = year, y = Abundance, fill = its2_type_profile, label = its2_type_profile)) +
    geom_bar(stat = "identity", position = "fill") +
    facet_wrap(~ host_site + host_id, nrow = 1) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = expansion(mult = c(0, .1)), labels = scales::percent) +
    ylab("Profiles") +
    scale_fill_manual(values = ppal[pmdf$its2_type_profile]) +
    # geom_label_repel(position = position_fill(vjust = 0.5), force = 2,
    #                  point.size = NA, xlim = c(-0.1, 3.6), ylim = c(0.1,0.9), size = 2) +
    theme(legend.position = "bottom", panel.background = element_blank(), 
          axis.title.x = element_blank(), axis.text.x = element_text(angle = 90),
          axis.text.y = element_blank(), axis.ticks.y = element_blank(),
          strip.background = element_blank(), strip.text.x = element_blank())# +
    #guides(fill = guide_legend(nrow = 4)) # +
    #coord_cartesian(clip = "off")
  plot_grid(v_plot, p_plot, nrow = 2, label_fontface = "plain")
}
```

```{r ggplot_theme}
# Create custom ggplot theme
theme_custom <- function() {
  theme_bw(base_size = 10, base_family = "Arial") %+replace%
    theme(
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(), 
      panel.background = element_blank(),
      panel.border = element_rect(color = "black", fill = NA),
      legend.background = element_rect(fill = NA, colour = NA)
    )
}

# scale_color_corals <- function(aesthetic, ...){
#     ggplot2:::manual_scale(
#         aesthetic,
#         values = setNames(c('#00BFC4', '#CD9600', '#7CAE00', '#00BE67', 
#                             '#F8766D', '#00A9FF', '#C77CFF', '#FF61CC'),
#                           c('annularis', 'cavernosa', 'faveolata', 'franksi', 
#                             'labrynthiformis', 'natans', 'siderea', 'strigosa')), 
#         ...
#     )
# }

scale_color_corals <- function(aesthetic, ...) {
  ggplot2:::manual_scale(
    aesthetic,
    labels = c("M. cavernosa", "S. siderea", "O. annularis", 
               "D. labyrinthiformis", "O. franksi", "C. natans", 
               "P. strigosa", "P. furcata", "O. faveolata"),
    values = setNames(c("#00B9E3", "#00BA38", "#F8766D",
                        "#93AA00", "#FF61C3", "#00C19F",
                        "#619CFF", "#D39200", "#DB72FB"),
                      c("cavernosa", "siderea", "annularis",
                        "labrynthiformis", "franksi", "natans",
                        "strigosa", "furcata", "faveolata")),
    ...)
}

sppord <- factor(levels = c("Cnat", "Dlab", "Pstr", "Pfur",
                            "Ssid", "Oann", "Ofav", "Ofra", "Mcav"))

## ggplot labeller
sppnames <- c(
  Cnat = "C. natans",
  Dlab = "D. labyrinthiformis",
  Pstr = "P. strigosa",
  Pfur = "P. furcata",
  Ssid = "S. siderea",
  Oann = "O. annularis",
  Ofav = "O. faveolata",
  Ofra = "O. franksi",
  Mcav = "M. cavernosa"
)

sitenames <- c(
  EastTektite = "ET",
  WhitePoint = "WP",
  CabritteHorn = "CH"
)

global_labeller <- labeller(
  gspp = sppnames,
  host_site = sitenames
)
```

```{r DAtesting_function}
# DESEq2 differential abundance testing function
deseqDA <- function(host_ids) {
  colData = data.frame(sample_data(variants))
  colData$host_id <- factor(colData$host_id)
  dds <- DESeqDataSetFromMatrix(countData = t(data.frame(otu_table(variants))),
                                colData = colData,
                                design = ~ year)
  ## Subset colonies
  dds <- dds[, colData(dds)$host_id %in% host_ids]
  ### Remove genes counted zero times in subset
  dds <- dds[ rowSums(counts(dds)) > 0, ]
  ### Drop unused factor levels
  colData(dds) <- droplevels(colData(dds))
  # Remove genes with less than 1 mean count across samples
  dds <- dds[ rowMeans(counts(dds)) > 1, ]
  # Run DESeq pipeline
  design(dds) <- formula(~ host_id + year)
  dsr1 <- DESeq(dds)
  # Define contrasts of interest (differences between years)
  contr <- tibble(num = c("2018", "2019", "2019"),
                  den = c("2017", "2017", "2018"))
  # Get diff abund taxa for each contrast
  DA <- contr %>%
    mutate(dsr = pmap(list(num, den), ~results(dsr1, contrast = c("year", ..1, ..2))))
  DA <- DA %>%
    mutate(sig = map(dsr, ~ rownames_to_column(data.frame(.[which(.$padj < 0.1), ]), "gene"))) %>%
    unnest(sig)
  # get variance stabilized counts for plotting
  vsd <- varianceStabilizingTransformation(dds)
  vsd2 <- limma::removeBatchEffect(assay(vsd), vsd$host_id)
  taxcounts <- data.frame(vsd2[which(rownames(vsd2) %in% DA$gene), ])
  if(ncol(taxcounts) == 1) {
    taxcounts <- data.frame(t(taxcounts))
    rownames(taxcounts) <- unique(DA$gene)
    }
  taxcounts <- rownames_to_column(taxcounts, var = "variant") %>%
    pivot_longer(-variant, names_to = "sample_name", values_to = "count") %>%
    left_join(dplyr::select(data.frame(sample_data(variants)), "sample_name", "host_id", "year"))
}



```

# Table of all sampling done
```{r}
# Create table of number of tagged colonies sampled at each site at each year
table1 <- as_tibble(sample_data(profiles)) %>%
  drop_na(host_id) %>%
  group_by(year, host_species, host_site) %>%
  summarize(n = n()) %>%
  arrange(host_site, year) %>%
  pivot_wider(names_from = c(host_site, year), values_from = n) %>% 
  adorn_totals(where = "row")

write_csv(table1, path = "Output/Table1.csv")
```

# QC (filter out low read samples)
```{r QC, include = FALSE}
# Total number of post-MED sequences
sum(sample_sums(variants))
nsamples(variants)

# Filter out samples with low read counts
low <- sample_sums(variants) < 1000
table(low)  # Number of samples being filtered out due to low read counts
variants <- subset_samples(variants, !low)
variants <- prune_taxa(taxa_sums(variants) != 0, variants)
profiles <- subset_samples(profiles, !low)
profiles <- prune_taxa(taxa_sums(profiles) != 0, profiles)

# Read count analysis
hist(sample_sums(variants), breaks = 100)
# One sample is an outlier with very high read count -- remove for summary stats of remaining samples
outlier <- sample_sums(variants)[which(sample_sums(variants) == max(sample_sums(variants)))]
woout <- sample_sums(variants)[which(sample_sums(variants) != max(sample_sums(variants)))]
# Histogram of remaining samples read counts
hist(woout, breaks = 50)
# Mean and standard deviation
mean(woout); sd(woout)

# Number of sequence variants and profiles
ntaxa(variants)
ntaxa(profiles)

# Parse sequences and profiles by genus(/clade)
data.frame(tax_table(variants)) %>% as_tibble() %>%
  count(clade)
data.frame(tax_table(profiles)) %>% as_tibble() %>%
  count(clade)
```

# Barplots for whole dataset
```{r}
# Algal genotypes --- whole dataset summary figure

## Transform to relative abundance
v <- transform_sample_counts(variants, function(x) x/sum(x))     # Relative abundance
p <- transform_sample_counts(profiles, function(x) x/sum(x))     # Relative abundance

## Filter out variants less than 0.1% relative abundance for plotting
vt <- transform_sample_counts(v, function(x) ifelse(x > 0.001, x, 0))
vt <- prune_taxa(taxa_sums(vt) != 0, vt)

### Arrange colonies by similarity within species
mp <- merge_samples(vt, "host_id")
hc <- hclust(dist(data.frame(otu_table(mp))))
neword <- tibble(host_id = as.character(sample_data(mp)$host_id), ord = hc$order)


## Set colors for plotting variants, with distinct palettes by clade/genus
tt <- as_tibble(data.frame(tax_table(vt))) %>%
  arrange(clade, DIV)
nt <- tt %>% count(clade) %>% pivot_wider(names_from = clade, values_from = n)
acols <- colorRampPalette(brewer.pal(9, "Greys"))(nt$A)
bcols <- colorRampPalette(brewer.pal(9, "Greens"))(nt$B)
ccols <- colorRampPalette(brewer.pal(9, "Blues"))(nt$C)
dcols <- colorRampPalette(brewer.pal(4, "Oranges"))(nt$D)
set.seed(556)
vpal <- c(acols[shuffle(acols)], bcols[shuffle(bcols)], ccols[shuffle(ccols)], dcols[shuffle(dcols)])
vpal <- setNames(vpal, tt$DIV)

# Order data for plotting
divord <- as_tibble(data.frame(tax_table(vt))) %>%
  arrange(clade, DIV) %>%
  mutate(DIV = as_factor(DIV))

  
vmdf <- psmelt(vt) %>%
  left_join(neword) %>%
  arrange(host_genus, host_species, host_site, ord, year, clade, DIV) %>%
  mutate(host_id = as_factor(host_id),
         sample_id = as_factor(paste(host_genus, host_species, host_id, year, sep = "_")),
         genus_species = as_factor(paste(str_sub(host_genus, 1, 1), ". ", host_species, sep = "")),
         gspp = factor(paste0(str_sub(host_genus, 1, 1), str_sub(host_species, 1, 3)),
                       levels = levels(sppord)),
         DIV = factor(DIV, levels = divord$DIV))

# Plot variants
g1 <- ggplot(vmdf, aes(x = year, y = Abundance)) + 
  geom_col(width = 1, aes(fill = DIV)) + 
  facet_nested(~ gspp + host_site + host_id, 
               labeller = global_labeller,
               space = "free_x",
               nest_line = element_line(linetype = 1),
               remove_labels = "x") +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0), labels = scales::percent) +
  ylab("ITS2 sequences") +
  scale_fill_manual(values = vpal[vmdf$OTU]) +
  theme(legend.position = "none", panel.background = element_blank(),
        text = element_text(size = 6), plot.margin = margin(5.5, 5.5, 1, 5.5),
        strip.text.x = element_text(margin = margin(0.5, 0, 0.5, 0, "mm"), face = "italic"),
        axis.title.x = element_blank(), axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.spacing = unit(0.25, "mm"),
        strip.background = element_blank())
g1

# Repeat for profiles
### truncate its2_type_profile names
pp <- as_tibble(data.frame(tax_table(p))) %>%
  arrange(clade, its2_type_profile) %>%
  mutate(its2_type_profile_trunc = str_trunc(its2_type_profile, 22, "right"))
np <- pp %>% count(clade) %>% pivot_wider(names_from = clade, values_from = n)
acols <- brewer.pal(np$A, "Greys")
bcols <- colorRampPalette(brewer.pal(9, "YlGn"))(np$B)
ccols <- colorRampPalette(brewer.pal(6, "PuBu"))(np$C)
dcols <- brewer.pal(np$D, "Reds")
set.seed(556)
ppal <- c(acols, bcols, ccols, dcols)
ppal <- setNames(ppal, pp$its2_type_profile_trunc)

proford <- as_tibble(data.frame(tax_table(p))) %>%
  mutate(its2_type_profile_trunc = str_trunc(its2_type_profile, 22, "right")) %>%
  arrange(clade, its2_type_profile_trunc) %>%
  mutate(its2_type_profile = as_factor(its2_type_profile_trunc))
  
pmdf <- psmelt(p) %>%
  left_join(neword) %>%
  arrange(host_genus, host_species, host_site, ord, year, clade, its2_type_profile) %>%
  mutate(host_id = as_factor(host_id),
         its2_type_profile_trunc = str_trunc(its2_type_profile, 22, "right"),
         sample_id = as_factor(paste(host_genus, host_species, host_id, year, sep = "_")),
         genus_species = as_factor(paste(str_sub(host_genus, 1, 1), ". ", host_species, sep = "")),
         gspp = factor(paste0(str_sub(host_genus, 1, 1), str_sub(host_species, 1, 3)),
                       levels = levels(sppord))) %>%
  filter(Abundance > 0.01)# %>% 
  #complete(year, host_id)

pmdf <- pmdf %>%
  mutate(yearlab = case_when(host_id == "12" ~ year,
                             TRUE ~ ""))
  
g2 <- ggplot(pmdf, aes(x = year, y = Abundance)) +
    geom_col(width = 1, aes(fill = its2_type_profile_trunc)) +
    facet_nested(~ gspp + host_site + host_id) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) +
    ylab("Profiles") +
    scale_fill_manual(values = ppal[pmdf$its2_type_profile_trunc], limits = force, drop = TRUE) +
    theme(legend.position = "bottom", legend.key.size = unit(2, "mm"), legend.title = element_blank(),
          legend.margin = margin(1, 0, 0, 0), legend.box.margin = margin(-10, -10, 0, -10),
          legend.key = element_rect(fill = NA),
          plot.margin = margin(0, 5.5, 5.5, 5.5),
          text = element_text(size = 6), legend.text = element_text(size = 4.4),
          panel.background = element_blank(), strip.text.x = element_blank(),
          axis.title.x = element_blank(), axis.text.y = element_blank(), 
          axis.text.x = element_text(angle = 90, vjust = 0.5, size = 3),
          axis.ticks.x = element_line(size = 0.25),
          axis.ticks.y = element_blank(),
          strip.background = element_blank(),
          panel.spacing = unit(0.25, "mm")) +
    guides(fill = guide_legend(ncol = 4))
g2
g2leg <- get_legend(g2)
g2 <- g2 + theme(legend.position = "none")
g2

fig1 <- plot_grid(g1, g2, nrow = 2, label_fontface = "plain", rel_heights = c(0.6, 0.4))

fig1

ggsave(fig1, filename = "Output/Figure1.png", width = 183, height = 70, units = "mm")
```

```{r}
# how many diff dom profiles in each species?
pmdf %>%
  filter(Abundance > 0.5) %>%
  group_by(host_species) %>%
  distinct(its2_type_profile)

pmdf %>%
  filter(Abundance > 0.5) %>%
  group_by(host_species) %>%
  summarize(n_colonies = n_distinct(host_id),
            n_domprof = n_distinct(its2_type_profile)) %>%
  arrange(n_domprof / n_colonies)

pmdf %>%
  filter(Abundance > 0.5) %>%
  group_by(host_species, its2_type_profile) %>%
  summarize(cols = paste0(unique(host_id), collapse = ",")) %>%
  print(n = nrow(.))


###### NETWORKS of dom profiles
library(igraph)
#only one connection per unique colony
df2 <- as_tibble(pmdf) %>%
  filter(Abundance > 0.5) %>%
  select(its2_type_profile, host_species, host_id, year) %>%
  distinct(its2_type_profile, host_species, host_id) %>%
  group_by(its2_type_profile, host_species) %>%
  mutate(n = n(),
         weights = 1/n) %>%
  ungroup()

net <- graph_from_data_frame(df2, directed = FALSE)

V(net)$type <- if_else(V(net)$name %in% unique(pmdf$host_species), "coral", "sym")
V(net)$clade <- if_else(V(net)$type == "sym", str_sub(V(net)$name, 1, 1), NA_character_)
V(net)$shape <- c("square", "circle")[factor(V(net)$type)]
taxcolors <- matrix(c("#f2f2f2", "#ccebc5", "#b3cde3", "#fbb4ae"), 
                    dimnames=list(c("A", "B", "C", "D")))
ppal <- c(acols, bcols, ccols, dcols)
ppal <- setNames(ppal, pp$its2_type_profile)
V(net)$color <- if_else(is.na(V(net)$clade), "white", taxcolors[factor(V(net)$clade, levels=c("A","B","C","D","G"))])
V(net)$color <- if_else(is.na(V(net)$clade), "white", ppal[V(net)$name])

 

V(net)$label <- sapply(V(net)$name, function(x) paste0(stri_wrap(str = x, width = 12, whitespace_only = FALSE),
          collapse = "\n"))
V(net)$label[29:37] <- c("Cnat", "Dlab", "Mcav", "Oann", "Ofav", "Ofra", "Pfur", "Pstr", "Ssid")
V(net)$label.cex = if_else(V(net)$type == "coral", 0.75, 0.4)
V(net)$label.color = "black"
V(net)$size = if_else(V(net)$type == "coral", 14, 20)


# set.seed(27)
# test.layout <- layout_(net, with_dh(weight.edge.lengths = edge_density(net)/100000))
# plot(net, layout = test.layout)

set.seed(1)
plot(net, layout = layout_with_fr(net, weights = E(net)$weights^1))

png("Output/Figure2.png", width = 180, height = 180, units = "mm", res = 300)
set.seed(1)
plot(net, layout = layout_with_fr(net, weights = E(net)$weights^1))
dev.off()


library(ggplotify)
mynet <- base2grob(~plot(net, layout = layout_with_fr(net, weights = E(net)$weights^1)))

bottom <- plot_grid(g2leg, mynet)
bigfig <- plot_grid(fig1, bottom, ncol = 1)
bigfig
```


# Summary of number of corals that changed dominant symbiont genera or profiles, or number of genera
```{r, }
# which profiles in which species?
favi <- pmdf %>%
  filter(host_family == "Faviidae") %>%
  select(Sample, gspp, Abundance, clade, its2_type_profile) %>%
  group_by(gspp, clade, its2_type_profile) %>%
  summarize(meanAbund = mean(Abundance), n = n()) %>%
  ungroup()
# 
favi %>%
  filter(clade == "B") %>%
  arrange(its2_type_profile) %>%
  arrange(-n) %>%
  print(n = nrow(.))
# 
# vmdf %>% filter(Abundance > 0) %>% filter(DIV == "B14b")
# 
# # Dlab's Breviolum is also found at in one Mcav sample...
# pmdf %>% filter(its2_type_profile == "B1-B1al-B1aj-B1ak-B1aa")
# # Dlab's other Breviolum is also found at in one Mcav sample...
# pmdf %>% filter(its2_type_profile == "B1-B1z-B1aa-B1ad-B1ac-B1ab")
# # Pstr's Breviolum is also found in Mcav, Ofav, Ofra, Ssid
# pmdf %>% filter(its2_type_profile == "B1-B1x-B1t-B14b-B1af-B1ae-B1ag")
# # Cnat's Breviolum is also found in Mcav and Ofra
# pmdf %>% filter(its2_type_profile == "B19/B5af-B19e-B19c-B5ag-B19d-B19f-B40a-B19h")
```



```{r, dom_bk_etc}
# Changes in dominant genus
genus_change <- pmdf %>%
  filter(Abundance >= 0.5) %>%
  group_by(gspp, host_id) %>%
  summarize(n_dom_clades = n_distinct(host_id, clade)) %>%
  filter(n_dom_clades > 1)

# Changes in dominant profile
prof_change <- pmdf %>%
  filter(Abundance >= 0.5) %>%
  group_by(gspp, clade, host_id) %>%
  summarize(n_dom_profs = n_distinct(host_id, clade, its2_type_profile)) %>%
  filter(n_dom_profs > 1)
prof_change %>% distinct(gspp)
# 13 colonies changed dominant profile within genus
# how many total colonies sampled more than one year?
pmdf %>%
  distinct(host_id, year) %>%
  count(host_id) %>%
  count(n > 1)
# 56 colonies sampled more than once, 4 shuffled genera, and 13 shuffled profiles, so 39 did not

####### new organization of all of this

# get abundance of each clade in each sample
clades <- vmdf %>%
  # Get clade level relative abundances
  group_by(gspp, host_site, host_id, year, clade) %>%
  summarize(Abundance = sum(Abundance)) %>%
  group_by(gspp, host_site, host_id, year) %>%
  pivot_wider(names_from = clade, values_from = Abundance) %>%
  ungroup()

# get number of clades over threshold abundance in each sample
nclades <- clades %>%
  nest(data = c(A, B, C, D)) %>%
  mutate(nclades.0.1 = map_dbl(data, ~ sum(. > 0.001)),
         nclades.1   = map_dbl(data, ~ sum(. > 0.01)),
         nclades.10  = map_dbl(data, ~ sum(. > 0.1)))

# how many colonies ever had multiple clades (all time points)?
mult <- nclades %>%
  group_by(gspp, host_site, host_id) %>%
  summarize(mult.0.1 = any(nclades.0.1 > 1),
            mult.1   = any(nclades.1 > 1),
            mult.10  = any(nclades.10 > 1))
table(mult$mult.0.1) # 64/70 (91%) colonies had multiple clades > 0.1% abund
table(mult$mult.1) # 33/70 (47%) colonies had multiple clades > 1% abund
table(mult$mult.10) # 11/70 (16%) colonies had multiple clades > 10% abund


```

# NMDS comparing hosts for each symbiont genus
```{r nmds_Fig2}
samples_clades <- vmdf %>%
  # Get clade level relative abundances
  group_by(Sample, gspp, host_site, host_id, year, clade) %>%
  summarize(Abundance = sum(Abundance)) %>%
  group_by(Sample, gspp, host_site, host_id, year) %>%
  ungroup()
samples_with_5pA <- samples_clades %>% filter(clade == "A", Abundance >= 0.05)
samples_with_5pB <- samples_clades %>% filter(clade == "B", Abundance >= 0.05)
samples_with_5pC <- samples_clades %>% filter(clade == "C", Abundance >= 0.05)
samples_with_5pD <- samples_clades %>% filter(clade == "D", Abundance >= 0.05)
#NMDS
# Look at just Symbiodinium
varA <- subset_samples(variants, sample_name %in% samples_with_5pA$Sample)
varA <- subset_taxa(varA, clade == "A")
# Run NMDS for visualization
set.seed(54)
varA.nmds <- ordinate(varA, "NMDS", "bray")
varA.nmds.s <- plot_ordination(varA, varA.nmds, type="samples")
# Look at just Breviolum
varB <- subset_samples(variants, sample_name %in% samples_with_5pB$Sample)
varB <- subset_taxa(varB, clade == "B")
# Run NMDS for visualization
set.seed(54)
varB.nmds <- ordinate(varB, "NMDS", "bray")
varB.nmds.s <- plot_ordination(varB, varB.nmds, type="samples")
# Look at just Cladocopium
varC <- subset_samples(variants, sample_name %in% samples_with_5pC$Sample)
varC <- subset_taxa(varC, clade == "C")
# Run NMDS for visualization
set.seed(55)
varC.nmds <- ordinate(varC, "NMDS", "bray")
varC.nmds.s <- plot_ordination(varC, varC.nmds, type="samples")
# Look at just Durusdinium corals
varD <- subset_samples(variants, sample_name %in% samples_with_5pD$Sample)
varD <- subset_taxa(varD, clade == "D")
# Run NMDS for visualization
set.seed(66)
varD.nmds <- ordinate(varD, "NMDS", "bray")
varD.nmds.s <- plot_ordination(varD, varD.nmds, type="samples")

# Plot NMDS for each genus
clade_nmds <- bind_rows(
  Symbiodinium = varA.nmds.s$data,
  Breviolum = varB.nmds.s$data, 
  Cladocopium = varC.nmds.s$data, 
  Durusdinium = varD.nmds.s$data,
  .id = "sym_genus"
) %>%
  arrange(host_id, year) %>%
  mutate(sym_genus = factor(sym_genus, levels = c("Symbiodinium", "Breviolum", 
                                                  "Cladocopium", "Durusdinium"))) %>%
  ggplot(aes(x = NMDS1, y = NMDS2, color = host_species, shape = host_species)) +
  facet_wrap(~ sym_genus, scales = "free", nrow = 1) +
  geom_point(alpha = 1, size = 2) +
  geom_path(aes(group = host_id), linewidth = 0.2) +
  scale_shape_manual(values = 8:0) +
  theme_custom() +
  theme(legend.position = "bottom",
        strip.background = element_blank(),
        strip.text = element_text(face = "italic"),
        legend.key.size = unit(2, "mm"), legend.title = element_blank(),
          legend.margin = margin(1, 0, 0, 0), legend.box.margin = margin(-10, -10, 0, -10),
          legend.key = element_rect(fill = NA),
          text = element_text(size = 6), legend.text = element_text(size = 4)) +
  guides(shape = guide_legend(ncol = 9))

clade_nmds
ggsave(clade_nmds, filename = "Output/Figure2.png", width = 183, height = 60, units = "mm")

clade_nmds +
  geom_point(aes(x = ))


as_tibble(varB.nmds.s$data) %>%
  filter(host_species == "siderea")

pmdf2 <- psmelt(p) 
pmdf2 %>% filter(host_id == "51", Abundance > 0.5)
pmdf2 %>% filter(its2_type_profile == "B1-B1z-B1aa-B1ad-B1ac-B1ab", Abundance > 0.5)
pmdf2 %>% filter(its2_type_profile == "B1-B1x-B1t-B14b-B1af-B1ae-B1ag", Abundance > 0.5)
pmdf2 %>% filter(host_species == "cavernosa", Abundance > 0.9)
pmdf2 %>% filter(grepl("C40g", its2_type_profile), Abundance > 0) %>% 
  as_tibble() %>% select(Sample, Abundance, host_id, year, host_species) %>% print(n = nrow(.))
pmdf2 %>% filter(grepl("C3an", its2_type_profile), Abundance > 0) %>% 
  as_tibble() %>% select(Sample, Abundance, host_id, year, host_species,
                         its2_type_profile) %>% print(n = nrow(.))
pmdf2 %>% filter(grepl("C7", its2_type_profile), Abundance > 0.01) %>% 
  as_tibble() %>% select(Sample, Abundance, host_id, year, host_species,
                         its2_type_profile) %>% print(n = nrow(.))
pmdf2 %>% filter(host_species == "strigosa", clade == "C", Abundance > 0.5)

pmdf2 %>% filter(grepl("C3fe", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(grepl("C54a", its2_type_profile), Abundance > 0)

pmdf2 %>% filter(host_species == "siderea", clade == "C", Abundance > 0.5)
pmdf2 %>% filter(grepl("C3a-", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(grepl("C16a", its2_type_profile), Abundance > 0)

pmdf2 %>% filter(host_species == "furcata", clade == "C", Abundance > 0)
pmdf2 %>% filter(grepl("C47", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(grepl("C42", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(grepl("C45a", its2_type_profile), Abundance > 0)


pmdf2 %>% filter(host_species == "natans", clade == "B", Abundance > 0)
pmdf2 %>% filter(grepl("B5af", its2_type_profile), Abundance > 0)
pmdf2 %>% filter(host_species == "strigosa", clade == "B", Abundance > 0)
pmdf2 %>% filter(grepl("B1x", its2_type_profile), Abundance > 0)

pmdf2 %>% filter(host_species == "franksi", clade == "B", Abundance > 0.05)
pmdf2 %>% filter(host_species == "siderea", clade == "B", Abundance > 0.05)
pmdf2 %>% filter(its2_type_profile == "B1-B1x-B1t-B14b-B1af-B1ae-B1ag", Abundance > 0.05)
pmdf2 %>% filter(its2_type_profile == "B19/B5af-B19e-B19c-B5ag-B19d-B19f-B40a-B19h", Abundance > 0.05)
pmdf2 %>% filter(its2_type_profile == "B1-B1z-B1aa-B1ad-B1ac-B1ab", Abundance > 0.05)
```

# Stats to compare symbionts within genus across hosts
```{r}
devtools::install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")

# Pairwise PERMANOVAs to test if the Breviolum sequence variants differ among host species

# Symbiodinium
Amod <- pairwiseAdonis::pairwise.adonis(distance(varA, "bray"), 
                                        as_tibble(sample_data(varA))$host_species,
                                        p.adjust.m = "bonferroni")
Amod

Amod <- Amod %>%
  separate(col = pairs, into = c("var1", "var2"), sep = " vs ")

Amod %>%
  bind_rows(Amod %>% rename(var1 = var2, var2 = var1)) %>%
  full_join(expand_grid(var1 = unique(as_tibble(sample_data(varA))$host_species), 
                        var2 = unique(as_tibble(sample_data(varA))$host_species))) %>%
  mutate(p.sig = case_when(p.adjusted < 0.05 ~ p.adjusted, TRUE ~ NA_real_)) %>%
  ggplot(aes(x = var1, y = var2)) +
  geom_tile(col = "black", aes(fill = log10(p.adjusted))) +
  geom_text(aes(label = sig), fontface = "bold", size = 5) +
  scale_fill_gradient(low = "red", high = "white")


# Breviolum
Bmod <- pairwiseAdonis::pairwise.adonis(distance(varB, "bray"), 
                                        as_tibble(sample_data(varB))$host_species,
                                        p.adjust.m = "fdr")
Bmod

Bmod <- Bmod %>%
  separate(col = pairs, into = c("var1", "var2"), sep = " vs ")

Bmod %>%
  bind_rows(Bmod %>% rename(var1 = var2, var2 = var1)) %>%
  full_join(expand_grid(var1 = unique(as_tibble(sample_data(varB))$host_species), 
                        var2 = unique(as_tibble(sample_data(varB))$host_species))) %>%
  mutate(p.sig = case_when(p.adjusted < 0.05 ~ p.adjusted, TRUE ~ NA_real_)) %>%
  ggplot(aes(x = var1, y = var2)) +
  geom_tile(col = "black", aes(fill = log10(p.adjusted))) +
  geom_text(aes(label = sig), fontface = "bold", size = 5) +
  scale_fill_gradient(low = "red", high = "white")

# Cladocopium
as_tibble(sample_data(varC)) %>% count(host_species)
Cmod <- pairwiseAdonis::pairwise.adonis(distance(varC, "bray"), 
                                        as_tibble(sample_data(varC))$host_species,
                                        p.adjust.m = "fdr")
Cmod

Cmod <- Cmod %>%
  separate(col = pairs, into = c("var1", "var2"), sep = " vs ")

Cmod %>%
  bind_rows(Cmod %>% rename(var1 = var2, var2 = var1)) %>%
  full_join(expand_grid(var1 = unique(as_tibble(sample_data(varC))$host_species), 
                        var2 = unique(as_tibble(sample_data(varC))$host_species))) %>%
  mutate(p.sig = case_when(p.adjusted < 0.05 ~ p.adjusted, TRUE ~ NA_real_)) %>%
  ggplot(aes(x = var1, y = var2)) +
  geom_tile(col = "black", aes(fill = log10(p.adjusted))) +
  geom_text(aes(label = sig), fontface = "bold", size = 5) +
  scale_fill_gradient(low = "red", high = "white")


# Durusdinium
as_tibble(sample_data(varD)) %>% count(host_species)
Dmod <- pairwiseAdonis::pairwise.adonis(distance(varD, "bray"), 
                                        as_tibble(sample_data(varD))$host_species,
                                        p.adjust.m = "fdr")
Dmod
```

# Stats to compare sites for each species
```{r}
## PERMANOVAs...
varr <- transform_sample_counts(variants, function(x) x/sum(x)) 


cnat <- subset_samples(varr, host_species == "natans")
adonis(distance(cnat, "bray") ~ host_site, data = as_tibble(sample_data(cnat)))
plot_bar(cnat, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

dlab <- subset_samples(varr, host_species == "labrynthiformis")
adonis(distance(dlab, "bray") ~ host_site, data = as_tibble(sample_data(dlab)))
plot_bar(dlab, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

pstr <- subset_samples(varr, host_species == "strigosa")
adonis(distance(pstr, "bray") ~ host_site, data = as_tibble(sample_data(pstr)))
plot_bar(pstr, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

mcav <- subset_samples(varr, host_species == "cavernosa")
adonis(distance(mcav, "bray") ~ host_site, data = as_tibble(sample_data(mcav)))
plot_bar(mcav, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

ssid <- subset_samples(varr, host_species == "siderea")
adonis(distance(ssid, "bray") ~ host_site, data = as_tibble(sample_data(ssid)))
plot_bar(ssid, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

ofav <- subset_samples(varr, host_species == "faveolata")
adonis(distance(ofav, "bray") ~ host_site, data = as_tibble(sample_data(ofav)))
plot_bar(ofav, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

oann <- subset_samples(varr, host_species == "annularis")
adonis(distance(oann, "bray") ~ host_site, data = as_tibble(sample_data(oann)))
plot_bar(oann, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

ofra <- subset_samples(varr, host_species == "franksi")
adonis(distance(ofra, "bray") ~ host_site, data = as_tibble(sample_data(ofra)))
plot_bar(ofra, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")

pfur <- subset_samples(varr, host_species == "furcata")
adonis(distance(pfur, "bray") ~ host_site, data = as_tibble(sample_data(pfur)))
plot_bar(pfur, x = "sample_name", fill = "clade") + facet_wrap(~host_site, scales = "free")
```


# How dynamic are coral symbiont communities within individual colonies? (change over time)
### Bray-Curtis diffs within colonies over time
```{r}


varr <- transform_sample_counts(variants, function(x) sqrt(x/sum(x))) 
#set.seed(55)
set.seed(56)
varr.nmds <- ordinate(varr, "NMDS", "bray")
varr.nmds.s <- plot_ordination(varr, varr.nmds, type="samples")

nmds_fig <- as_tibble(varr.nmds.s$data) %>%
  mutate(gspp = factor(paste0(str_sub(host_genus, 1, 1), str_sub(host_species, 1, 3)))) %>%
  arrange(host_id, year) %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(aes(shape = year, color = year), alpha = 0.6) +
  geom_path(aes(group = host_id), alpha = 0.5, arrow = arrow(ends = "last", length = unit(1, "mm")), lwd = 0.25) +
  facet_wrap(~gspp, labeller = global_labeller) +
  scale_color_manual(values = c('#66c2a5','#fc8d62','#8da0cb')) +
  theme_custom() +
  theme(strip.background = element_blank(),
          strip.text = element_text(face = "italic"),
          text = element_text(size = 7), legend.text = element_text(size = 4),
          axis.text.x = element_text(hjust = 0)) +
  theme(legend.position = c(0.1, 0.1)); nmds_fig

ggsave(nmds_fig, filename = "Output/Figure3.png", width = 85, height = 85, units = "mm")




# include as supplementary?
as_tibble(varr.nmds.s$data) %>%
  group_by(host_id) %>%
  filter(any(year == "2017")) %>%
  mutate(NMDS1adj = NMDS1 - NMDS1[year == "2017"],
         NMDS2adj = NMDS2 - NMDS2[year == "2017"]) %>%
  ungroup() %>%
  arrange(host_id, year) %>%
  ggplot(aes(x = NMDS1adj, y = NMDS2adj)) +
  geom_point(aes(shape = year, color = year)) +
  geom_path(aes(group = host_id), alpha = 0.5, arrow = arrow(ends = "last", length = unit(2, "mm"))) +
  facet_wrap(~host_species) +
  scale_x_continuous(limits = c(-1, 1)) +
  scale_y_continuous(limits = c(-1, 1)) +
  theme_custom()
```

# Compare between-colony and within-colony Bray-Curtis dissimilarities across species

```{r}
varr <- transform_sample_counts(variants, function(x) sqrt(x/sum(x)))    # Square-root transform

bd <- as.matrix(distance(varr, "bray"))
xy <- t(combn(colnames(bd), 2))
bd2 <- as_tibble(data.frame(xy, dist=bd[xy])) %>%
  separate(X1, into = c("site1", "species1", "host_id1", "year1")) %>%
  separate(X2, into = c("site2", "species2", "host_id2", "year2")) %>%
  filter(species1 == species2) %>%    # only interested in within-species differences
  mutate(group = case_when(host_id1 == host_id2 ~ "within", TRUE ~ "between")) %>%
  filter(group == "within" |
         group == "between" & year1 == year2)



#order species by highest median between-colony difference
brayord <- bd2 %>%
  filter(group == "between") %>%
  group_by(species1) %>%
  summarize(medbray = median(dist)) %>%
  arrange(-medbray) %>%
  mutate(gspp = as_factor(species1))
  
source("https://raw.githubusercontent.com/datavizpyr/data/master/half_flat_violinplot.R")
f4 <- bd2 %>% mutate(gspp = factor(species1, levels = levels(brayord$gspp))) %>%
  ggplot(aes(x = group, y = dist)) +
    facet_wrap(~ gspp, nrow = 2, labeller = global_labeller) +
    geom_flat_violin(adjust = 1, position = position_nudge(x = 0.15), scale = "width",
                     lwd = 0, fill = "grey72") +
    geom_boxplot(width = 0.1, outlier.shape = NA, lwd = 0.2,
                 position = position_nudge(x = 0.15)) +
    geom_jitter(width = 0.075, alpha = 0.2, stroke = 0) +
    theme_custom() +
    theme(strip.background = element_blank(),
          strip.text = element_text(face = "italic"),
          text = element_text(size = 7), legend.text = element_text(size = 4),,
          axis.text.x = element_text(hjust = 0)) +
    labs(y = "Bray-Curtis Dissimilarity", x = "") +
    coord_cartesian(xlim = c(1.3, 2.1)); f4

ggsave(f4, filename = "Output/Figure4.png", width = 102, height = 70, units = "mm")

bd2 %>%
  group_by(species1, group) %>%
  summarize(median_dist = median(dist)) %>%
  pivot_wider(names_from = group, values_from = median_dist) %>%
  mutate(btw_wit = between/within) %>%
  arrange(btw_wit)


```

